// tea-cup-core/dist/Cmd.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
/**
 * Base class for Commands.
 */
export class Cmd {
    /**
     * Create a command that does nothing
     */
    static none() {
        return new CmdNone();
    }
    /**
     * Batches passed commands into a single command.
     * @param cmds the commands to batch
     */
    static batch(cmds) {
        return new BatchCmd(cmds);
    }
    /**
     * Map this command, useful for parent-child scenarios.
     * @param mapper
     */
    map(mapper) {
        return new CmdMapped(this, mapper);
    }
}
/**
 * A command that does nothing.
 */
// exported for perf optimisation reasons
export class CmdNone extends Cmd {
    execute(dispatch) {
        // it's a noop !
    }
}
/**
 * Utility function for transforming an object into a
 * object/Cmd.none() tuple.
 * @param t an object to be paired with a Cmd.none().
 */
export function noCmd(t) {
    return [t, Cmd.none()];
}
class CmdMapped extends Cmd {
    constructor(command, mapper) {
        super();
        this.command = command;
        this.mapper = mapper;
    }
    execute(dispatch) {
        this.command.execute((m) => {
            dispatch(this.mapper(m));
        });
    }
}
class BatchCmd extends Cmd {
    constructor(cmds) {
        super();
        this.cmds = cmds;
    }
    execute(dispatch) {
        this.cmds.forEach((cmd) => cmd.execute(dispatch));
    }
}
