// tea-cup-core/dist/Decode.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { err, ok } from './Result';
import { just, nothing } from './Maybe';
import { List } from './List';
import { asError } from './Try';
/**
 * Decoder for a given type.
 */
export class Decoder {
    constructor(f) {
        this.f = f;
    }
    /**
     * Attempt to decode passed JSON string into a T
     * @param s the string to decode
     */
    decodeString(s) {
        try {
            const o = JSON.parse(s);
            return this.decodeValue(o);
        }
        catch (e) {
            return err(asError(e).message);
        }
    }
    /**
     * Attempt to decode a JS value into a T
     * @param o the value to decode
     */
    decodeValue(o) {
        return this.f(o);
    }
}
function stringifyForMsg(o, maxChars = 100) {
    if (o === null) {
        return 'null';
    }
    if (o === undefined) {
        return 'undefined';
    }
    try {
        const s = JSON.stringify(o);
        if (s.length > maxChars) {
            return s.substring(0, maxChars - 1);
        }
        else {
            return s;
        }
    }
    catch (e) {
        return o.toString();
    }
}
/**
 * Decoding primitives and utilities.
 */
export class Decode {
    // Data Structures
    /**
     * Decoder for null/undefined values
     * @param d the decoder to be used if the value is not null or undefined
     */
    static nullable(d) {
        return new Decoder((o) => {
            if (o === null || o === undefined) {
                return ok(nothing);
            }
            else {
                return d.decodeValue(o).map(just);
            }
        });
    }
    /**
     * Decoder for nullable types
     * @param d the decoder to be used if the value is not null
     */
    static orNull(d) {
        return this.map((v) => v.map((v) => v).withDefault(null), this.nullable(d));
    }
    /**
     * Decoder for lists
     * @param d the decoder for elements in the list
     */
    static list(d) {
        return Decode.map((a) => List.fromArray(a), Decode.array(d));
    }
    /**
     * Decoder for arrays
     * @param d the decoder for elements in the array
     */
    static array(d) {
        return new Decoder((o) => {
            if (o instanceof Array) {
                const a = o;
                const res = [];
                for (let i = 0; i < a.length; i++) {
                    const r = d.decodeValue(a[i]);
                    switch (r.tag) {
                        case 'Ok':
                            res.push(r.value);
                            break;
                        case 'Err':
                            return err(`could not convert element at index ${i} of ${stringifyForMsg(o)} : ${r.err}`);
                    }
                }
                return ok(res);
            }
            else {
                return err(`value is not an array : ${stringifyForMsg(o)}`);
            }
        });
    }
    // Object Primitives
    /**
     * Decoder for object fields
     * @param key the name of the field
     * @param d the decoder for the field's value
     */
    static field(key, d) {
        return Decode.at([key], d);
    }
    /**
     * Decoder for optinal object fields
     * @param key the name of the optinal field
     * @param d the decoder for the field's value
     */
    static optionalField(key, d) {
        return Decode.andThen((value) => value
            .map((v) => new Decoder(() => d.decodeValue(v)))
            .withDefault(Decode.succeed(undefined)), Decode.maybe(Decode.field(key, Decode.value)));
    }
    /**
     * Decoder for navigable object properties
     * @param keys a list of fields to navigate
     * @param d the decoder for the leaf value
     */
    static at(keys, d) {
        function isUndef(x) {
            return x === null || x === undefined;
        }
        return new Decoder((o) => {
            let v = o;
            const path = [];
            function pathToStr() {
                return `[${path.join(',')}]`;
            }
            for (let i = 0; i < keys.length; i++) {
                if (isUndef(v)) {
                    return err(`path not found ${pathToStr()} on ${stringifyForMsg(o)}`);
                }
                const key = keys[i];
                path.push(key);
                if (v.hasOwnProperty(key)) {
                    v = v[key];
                }
                else {
                    return err(`path not found ${pathToStr()} on ${stringifyForMsg(o)}`);
                }
            }
            const r = d.decodeValue(v);
            switch (r.tag) {
                case 'Ok':
                    return r;
                case 'Err':
                    return err(`ran into decoder error at ${pathToStr()} : ${r.err}`);
            }
        });
    }
    // Inconsistent Structure
    /**
     * Decoder for optional values : turns decoding failures into maybes.
     * @param d the decoder to be used
     */
    static maybe(d) {
        return new Decoder((o) => {
            const v = d.decodeValue(o);
            switch (v.tag) {
                case 'Ok':
                    return ok(just(v.value));
                case 'Err':
                    return ok(nothing);
            }
        });
    }
    /**
     * Tries passed decoders sequentially, and fails if no decoder succeeds
     * @param ds an array of Decoders to try, one after the othen
     */
    static oneOf(ds) {
        return new Decoder((o) => {
            for (let i = 0; i < ds.length; i++) {
                const r = ds[i].decodeValue(o);
                switch (r.tag) {
                    case 'Ok':
                        return r;
                    case 'Err':
                        break;
                }
            }
            return err(`ran out of decoders for ${stringifyForMsg(o)}`);
        });
    }
    // Mapping
    /**
     * Map a decoder
     * @param f the mapping function
     * @param d the decoder to use
     */
    static map(f, d) {
        return new Decoder((o) => {
            return d.decodeValue(o).map(f);
        });
    }
    static map2(f, d1, d2) {
        return Decode.andThen((t1) => Decode.andThen((t2) => {
            const t3 = f(t1, t2);
            return Decode.succeed(t3);
        }, d2), d1);
    }
    static map3(f, d1, d2, d3) {
        return Decode.andThen((t1) => Decode.map2((t2, t3) => {
            return f(t1, t2, t3);
        }, d2, d3), d1);
    }
    static map4(f, d1, d2, d3, d4) {
        return Decode.andThen((t1) => Decode.map3((t2, t3, t4) => {
            return f(t1, t2, t3, t4);
        }, d2, d3, d4), d1);
    }
    static map5(f, d1, d2, d3, d4, d5) {
        return Decode.andThen((t1) => Decode.map4((t2, t3, t4, t5) => {
            return f(t1, t2, t3, t4, t5);
        }, d2, d3, d4, d5), d1);
    }
    static map6(f, d1, d2, d3, d4, d5, d6) {
        return Decode.andThen((t1) => Decode.map5((t2, t3, t4, t5, t6) => {
            return f(t1, t2, t3, t4, t5, t6);
        }, d2, d3, d4, d5, d6), d1);
    }
    static map7(f, d1, d2, d3, d4, d5, d6, d7) {
        return Decode.andThen((t1) => Decode.map6((t2, t3, t4, t5, t6, t7) => {
            return f(t1, t2, t3, t4, t5, t6, t7);
        }, d2, d3, d4, d5, d6, d7), d1);
    }
    static map8(f, d1, d2, d3, d4, d5, d6, d7, d8) {
        return Decode.andThen((t1) => Decode.map7((t2, t3, t4, t5, t6, t7, t8) => {
            return f(t1, t2, t3, t4, t5, t6, t7, t8);
        }, d2, d3, d4, d5, d6, d7, d8), d1);
    }
    /**
     * Decoder for big objects, where map8() is not enough.
     * @param dobject an object with decoders
     */
    static mapObject(dobject) {
        const keys = Object.keys(dobject);
        return new Decoder((value) => keys.reduce((acc, key) => {
            const propertyDecoder = getProperty(dobject, key);
            const v = propertyDecoder.decodeValue(value);
            return v.andThen((v) => acc.map((acc) => {
                acc[key] = v;
                return acc;
            }));
        }, ok({})));
    }
    /**
     * Convenience, map decoder object to another decoder object
     * @param decoders an object with decoders
     * @param fun the mapper function
     */
    static mapFields(decoders, fun) {
        const keys = Object.keys(decoders);
        const partial = keys.reduce((acc, key) => {
            const propertyDecoder = getProperty(decoders, key);
            const [key2, propertyDecoder2] = fun(key, propertyDecoder);
            acc[key2] = propertyDecoder2;
            return acc;
        }, {});
        return partial;
    }
    /**
     * Convenience, map docoders to required field decoders
     * @param decoders an object with decoders
     */
    static mapRequiredFields(decoders) {
        return this.mapFields(decoders, (k, d) => [k, Decode.field(k, d)]);
    }
    /**
     * Convenience, map decoders to optional field decoders
     * @param decoders an object with decoders
     */
    static mapOptionalFields(decoders) {
        const mapper = (k, d) => [
            k,
            Decode.optionalField(k, d),
        ];
        return this.mapFields(decoders, mapper);
    }
    /**
     * Decoder for fixed-length tuples.
     * @param decoders an array with decoders
     */
    static mapTuple(decoders) {
        return Decode.map((v) => Object.values(v), this.mapObject(this.mapRequiredFields(decoders)));
    }
    // Fancy Decoding
    /**
     * Decoder for recursive data structures
     * @param f a no-arg function that yields a decoder
     */
    static lazy(f) {
        return new Decoder((o) => {
            return f().decodeValue(o);
        });
    }
    /**
     * Decoder for null
     * @param t the result to yield in case the decoded value is null
     */
    static null(t) {
        return new Decoder((o) => {
            if (o === null) {
                return ok(t);
            }
            else {
                return err(`expected null for ${stringifyForMsg(o)}`);
            }
        });
    }
    /**
     * Decoder that always succeed
     * @param t the value to yield
     */
    static succeed(t) {
        return new Decoder(() => {
            return ok(t);
        });
    }
    /**
     * Decoder that fails
     * @param msg the message to use in the resulting Err
     */
    static fail(msg) {
        return new Decoder(() => {
            return err(msg);
        });
    }
    /**
     * Chain decoders
     * @param f the function to apply if the first decoder has succeeded
     * @param d the first decoder to use
     */
    static andThen(f, d) {
        return new Decoder((o) => {
            const r = d.decodeValue(o);
            switch (r.tag) {
                case 'Ok':
                    const t1 = r.value;
                    const d2 = f(t1);
                    return d2.decodeValue(o);
                case 'Err':
                    return err(r.err);
            }
        });
    }
}
// Primitives
/**
 * string decoder
 */
Decode.str = new Decoder((o) => {
    if (o !== null && o !== undefined && typeof o === 'string') {
        return ok(o);
    }
    else {
        return err(`value is not a string : ${stringifyForMsg(o)}`);
    }
});
/**
 * boolean decoder
 */
Decode.bool = new Decoder((o) => {
    if (o !== null && o !== undefined && typeof o === 'boolean') {
        return ok(o);
    }
    else {
        return err(`value is not a boolean : ${stringifyForMsg(o)}`);
    }
});
/**
 * number decoder
 */
Decode.num = new Decoder((o) => {
    if (o !== null && o !== undefined && typeof o === 'number') {
        return ok(o);
    }
    else {
        return err(`value is not a number : ${stringifyForMsg(o)}`);
    }
});
/**
 * Decoder for any value
 */
Decode.value = new Decoder((o) => ok(o));
function getProperty(o, key) {
    return o[key];
}
