// tea-cup-core/dist/Dict.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { Tuple } from './Tuple';
import { maybeOf } from './Maybe';
export class Dict {
    constructor(d) {
        this.d = d;
    }
    static empty() {
        return new Dict({});
    }
    static fromList(entries) {
        const newD = {};
        entries.forEach((e) => {
            newD[e.a] = e.b;
        });
        return new Dict(newD);
    }
    put(key, value) {
        const newD = Object.assign({}, this.d);
        newD[key] = value;
        return new Dict(newD);
    }
    get(key) {
        return maybeOf(this.d[key]);
    }
    exists(key) {
        return this.d[key] !== undefined;
    }
    remove(key) {
        if (this.d[key]) {
            const newD = Object.assign({}, this.d);
            delete newD[key];
            return new Dict(newD);
        }
        return this;
    }
    keys() {
        return Object.keys(this.d);
    }
    size() {
        return this.keys().length;
    }
    isEmpty() {
        return this.size() === 0;
    }
    toList() {
        return this.keys().map((k) => new Tuple(k, this.d[k]));
    }
    filter(f) {
        return Dict.fromList(this.toList().filter(f));
    }
    update(key, f) {
        return f(this.get(key))
            .map((value) => this.put(key, value))
            .withDefaultSupply(() => this.remove(key));
    }
}
