// tea-cup-core/dist/Either.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { just, nothing } from './Maybe';
export class Left {
    constructor(value) {
        this.tag = 'Left';
        this.value = value;
    }
    isLeft() {
        return true;
    }
    isRight() {
        return !this.isLeft();
    }
    mapLeft(f) {
        return new Left(f(this.value));
    }
    mapRight(f) {
        return new Left(this.value);
    }
    match(onLeft, onRight) {
        return onLeft(this.value);
    }
    get left() {
        return just(this.value);
    }
    get right() {
        return nothing;
    }
}
export class Right {
    constructor(value) {
        this.tag = 'Right';
        this.value = value;
    }
    isLeft() {
        return false;
    }
    isRight() {
        return !this.isLeft();
    }
    mapLeft(f) {
        return new Right(this.value);
    }
    mapRight(f) {
        return new Right(f(this.value));
    }
    match(onLeft, onRight) {
        return onRight(this.value);
    }
    get left() {
        return nothing;
    }
    get right() {
        return just(this.value);
    }
}
export function left(a) {
    return new Left(a);
}
export function right(b) {
    return new Right(b);
}
