// tea-cup-core/dist/Lens.js
/*
 * MIT License
 *
 * Copyright (c) 2024 Rémi Van Keisbelck, Frank Wagner
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { just, nothing } from "./Maybe";
export class Lens {
    constructor(get, set) {
        this.get = get;
        this.set = set;
    }
    update(a, f) {
        return this.set(a, f(this.get(a)));
    }
    updateWithResult(a, f) {
        const [b, r] = f(this.get(a));
        return [this.set(a, b), r];
    }
    field(field) {
        return composeLenses(this, fieldLens(field));
    }
    sub(f) {
        return composeLensPrism(this, subPrism(f));
    }
    flatten() {
        return composeLensPrism(this, flattenPrism());
    }
    andThenLens(lens) {
        return composeLenses(this, lens);
    }
    andThenPrism(prism) {
        return composeLensPrism(this, prism);
    }
}
export class Prism {
    constructor(get, set) {
        this.get = get;
        this.set = set;
    }
    update(a, f) {
        return this.get(a)
            .map(f)
            .map((fb) => this.set(a, fb));
    }
    updateWithResult(a, f) {
        return this.get(a)
            .map(f)
            .map(([b, r]) => [this.set(a, b), r]);
    }
    field(field) {
        return composePrismLens(this, fieldLens(field));
    }
    sub(f) {
        return composePrisms(this, subPrism(f));
    }
    andThen(lens) {
        if (lens instanceof Lens) {
            return composePrismLens(this, lens);
        }
        else {
            return composePrisms(this, lens);
        }
    }
    flatten() {
        return composePrisms(this, flattenPrism());
    }
}
export function idLens() {
    return new Lens((a) => a, (_, v) => v);
}
export function composeLenses(l1, l2) {
    return new Lens((a) => l2.get(l1.get(a)), (a, v) => l1.set(a, l2.set(l1.get(a), v)));
}
export function composeLensPrism(l, p) {
    return new Prism((a) => p.get(l.get(a)), (a, v) => l.set(a, p.set(l.get(a), v)));
}
export function composePrismLens(p, l) {
    return new Prism((a) => p.get(a).map(l.get), (a, v) => p
        .get(a)
        .map((b) => l.set(b, v))
        .map((bc) => p.set(a, bc))
        .withDefault(a));
}
export function composePrisms(p, p2) {
    return new Prism((a) => p.get(a).andThen(p2.get), (a, v) => p
        .get(a)
        .map((b) => p2.set(b, v))
        .map((bc) => p.set(a, bc))
        .withDefault(a));
}
export function fieldLens(field) {
    return new Lens((a) => a[field], (a, v) => {
        const a1 = Object.assign({}, a);
        a1[field] = v;
        return a1;
    });
}
export function flattenPrism() {
    return new Prism((a) => a, (a, v) => a.map(() => v).orElse(a));
}
export function discriminate(tag, v) {
    return (o) => (o[tag] === v ? just(o) : nothing);
}
export function subPrism(f) {
    return new Prism((a) => f(a), (a, v) => f(a)
        .map(() => v)
        .withDefault(a));
}
