// tea-cup-core/dist/List.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { just, nothing } from './Maybe';
/**
 * Immutable list of data
 */
export class List {
    constructor(elems) {
        this.elems = [...elems];
    }
    static fromArray(ts) {
        if (ts.length === 0) {
            return List.empty();
        }
        else {
            return new List(ts);
        }
    }
    static empty() {
        return new List([]);
    }
    head() {
        if (this.length() === 0) {
            return nothing;
        }
        else {
            return just(this.elems[0]);
        }
    }
    length() {
        return this.elems.length;
    }
    tail() {
        if (this.length() <= 1) {
            return new List([]);
        }
        else {
            const tailArray = [...this.elems];
            tailArray.shift();
            return new List(tailArray);
        }
    }
    toArray() {
        return [...this.elems];
    }
    map(f) {
        return new List(this.elems.map(f));
    }
    mapWithIndex(f) {
        return new List(this.elems.map(f));
    }
}
