// tea-cup-core/dist/Maybe.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
/**
 * A Maybe with something inside
 */
export class Just {
    constructor(value) {
        this.type = 'Just';
        this.value = value;
    }
    map(f) {
        return new Just(f(this.value));
    }
    withDefault(_) {
        return this.value;
    }
    withDefaultSupply(f) {
        return this.value;
    }
    orElse(e) {
        return this;
    }
    orElseSupply(f) {
        return this;
    }
    forEach(f) {
        f(this.value);
    }
    andThen(f) {
        return f(this.value);
    }
    toNative() {
        return this.value;
    }
    isJust() {
        return true;
    }
    isNothing() {
        return false;
    }
    filter(f) {
        if (f(this.value)) {
            return this;
        }
        return nothing;
    }
}
/**
 * Wrap an object into a Maybe
 * @param t the object to wrap
 */
export function just(t) {
    return new Just(t);
}
/**
 * An "empty" Maybe
 */
export class Nothing {
    constructor() {
        this.type = 'Nothing';
    }
    map(f) {
        return Nothing.value;
    }
    withDefault(d) {
        return d;
    }
    withDefaultSupply(f) {
        return f();
    }
    orElse(e) {
        return e;
    }
    orElseSupply(f) {
        return f();
    }
    forEach(f) { }
    andThen(f) {
        return nothing;
    }
    toNative() {
        return undefined;
    }
    isJust() {
        return false;
    }
    isNothing() {
        return true;
    }
    filter() {
        return nothing;
    }
}
Nothing.value = new Nothing();
/**
 * The "nothing" maybe.
 */
export const nothing = Nothing.value;
/**
 * Make a Maybe with a "nullable" object
 * @param t the object, or undef, or null
 */
export function maybeOf(t) {
    if (t === undefined || t === null) {
        return nothing;
    }
    else {
        return just(t);
    }
}
/**
 * Apply a function if all the arguments are `Just` a value.
 * @param m1 First Maybe
 * @param m2 Second Maybe
 * @param func the function applied on the values of the Maybes
 */
export function map2(m1, m2, func) {
    return m1
        .map((a) => {
        return m2
            .map((b) => {
            return new Just(func(a, b));
        })
            .withDefault(nothing);
    })
        .withDefault(nothing);
}
