// tea-cup-core/dist/Port.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { Sub } from './Sub';
export class Port {
    constructor() {
        this.subs = [];
    }
    send(t) {
        this.subs.forEach((s) => s.notify(t));
    }
    subscribe(f) {
        return new PortSub(f, (p) => this.subs.push(p), (p) => {
            this.subs = this.subs.filter((x) => x !== p);
        });
    }
}
class PortSub extends Sub {
    constructor(f, _onInit, _onRelease) {
        super();
        this.f = f;
        this._onInit = _onInit;
        this._onRelease = _onRelease;
    }
    onInit() {
        this._onInit(this);
    }
    onRelease() {
        this._onRelease(this);
    }
    notify(t) {
        this.dispatch(this.f(t));
    }
}
