// tea-cup-core/dist/Result.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { just, nothing } from './Maybe';
export class Ok {
    constructor(value) {
        this.tag = 'Ok';
        this.value = value;
    }
    map(f) {
        return new Ok(f(this.value));
    }
    mapError(f) {
        return new Ok(this.value);
    }
    toMaybe() {
        return just(this.value);
    }
    match(onOk, onErr) {
        return onOk(this.value);
    }
    withDefault(v) {
        return this.value;
    }
    withDefaultSupply(f) {
        return this.value;
    }
    andThen(f) {
        return f(this.value);
    }
    orElse(f) {
        return this;
    }
}
export class Err {
    constructor(err) {
        this.tag = 'Err';
        this.err = err;
    }
    map(f) {
        return new Err(this.err);
    }
    mapError(f) {
        return new Err(f(this.err));
    }
    toMaybe() {
        return nothing;
    }
    match(onOk, onErr) {
        return onErr(this.err);
    }
    withDefault(r) {
        return r;
    }
    withDefaultSupply(f) {
        return f();
    }
    andThen(f) {
        return err(this.err);
    }
    orElse(f) {
        return f(this.err);
    }
}
/**
 * Create an Ok Result
 * @param r the value to wrap
 */
export function ok(r) {
    return new Ok(r);
}
/**
 * Create an Err result
 * @param e the error to wrap
 */
export function err(e) {
    return new Err(e);
}
