// tea-cup-core/dist/Sub.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
export class Sub {
    constructor() {
        this.active = false;
    }
    static none() {
        return new SubNone();
    }
    static batch(subs) {
        return new BatchSub(subs);
    }
    init(dispatch) {
        this.dispatcher = dispatch;
        this.active = true;
        this.onInit();
    }
    release() {
        this.active = false;
        this.onRelease();
    }
    isActive() {
        return this.active;
    }
    dispatch(m) {
        var _a;
        (_a = this.dispatcher) === null || _a === void 0 ? void 0 : _a.call(this, m);
    }
    onInit() { }
    onRelease() { }
    map(f) {
        return new MappedSub(this, f);
    }
}
class SubNone extends Sub {
}
class BatchSub extends Sub {
    constructor(subs) {
        super();
        this.subs = subs;
    }
    init(dispatch) {
        this.subs.forEach((s) => s.init(dispatch));
    }
    release() {
        this.subs.forEach((s) => s.release());
    }
}
class MappedSub extends Sub {
    constructor(childSub, mapper) {
        super();
        this.childSub = childSub;
        this.mapper = mapper;
    }
    init(dispatch) {
        this.childSub.init((m) => {
            dispatch(this.mapper(m));
        });
    }
    release() {
        this.childSub.release();
    }
}
