// tea-cup-core/dist/Task.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { Cmd } from './Cmd';
import { err, Err, ok } from './Result';
import { just, nothing } from './Maybe';
import { asError } from './Try';
/**
 * Base class for Tasks.
 */
export class Task {
    /**
     * Send a Task to the runtime for execution
     * @param t the task
     * @param toMsg a function that turns the result of the task into a Msg
     */
    static attempt(t, toMsg) {
        return new TaskCmd(t, toMsg);
    }
    /**
     * Send a Task that never fails to the runtime for execution
     * @param t the task
     * @param toMsg a function that turns the result of the task into a Msg
     */
    static perform(t, toMsg) {
        return new TaskNoErrCmd(t, toMsg);
    }
    /**
     * Create a task that succeeds with a value
     * @param r the value
     */
    static succeed(r) {
        return new TSuccess(() => r);
    }
    /**
     * Create a task that succeeds with a lazily-supplied value
     * @param r the value supplier
     */
    static succeedLazy(r) {
        return new TSuccess(r);
    }
    /**
     * Create a task that fails with an error
     * @param e the error
     */
    static fail(e) {
        return new TError(() => e);
    }
    /**
     * Create a task that fails with a lazily-supplied error
     * @param e the error supplier
     */
    static failLazy(e) {
        return new TError(e);
    }
    /**
     * Create a task from a Promise
     * @param promiseSupplier a function that returns the promise (will be called on Task execution)
     */
    static fromPromise(promiseSupplier) {
        return new TPromise(promiseSupplier);
    }
    /**
     * Create a task from a supplier (lazy). Will return an error if the lambda throws an error.
     */
    static fromLambda(lambda) {
        return new TLambda(lambda);
    }
    /**
     * Map the ok result of this task
     * @param f the mapping function
     */
    map(f) {
        return new TMapped(this, f);
    }
    /**
     * Map the error result of this task
     * @param f the mapping function
     */
    mapError(f) {
        return new TMappedErr(this, f);
    }
    /**
     * Recover from error (turns error into a success !)
     * @param f the error-to-success converter
     */
    recover(f) {
        return new TRecover(this, f);
    }
    /**
     * Chain this task with another task
     * @param f a function that accepts the result of this task, and yields a new task
     */
    andThen(f) {
        return new TThen(this, f);
    }
    /**
     * Runs tasks in parallel
     * @param t the task to be run in parallel with this task
     * @param f a function that maps the results of the 2 parallel tasks
     */
    parallel(t, f) {
        return new TParallel(f, this, t);
    }
}
class TRecover extends Task {
    constructor(t, f) {
        super();
        this.t = t;
        this.f = f;
    }
    execute(callback) {
        this.t.execute((tRes) => tRes.match((tOk) => callback(ok(tOk)), (tErr) => callback(ok(this.f(tErr)))));
    }
}
class TParallel extends Task {
    constructor(f, t1, t2) {
        super();
        this.f = f;
        this.t1 = t1;
        this.t2 = t2;
    }
    execute(callback) {
        let ra = nothing;
        let rb = nothing;
        let error = nothing;
        const done = () => {
            if (error.type !== 'Nothing' || ra.type === 'Nothing' || rb.type === 'Nothing') {
                return;
            }
            callback(ok(this.f(ra.value, rb.value)));
        };
        function handle(t, assign) {
            t.execute((r) => {
                switch (r.tag) {
                    case 'Err': {
                        if (error === null || error === void 0 ? void 0 : error.isNothing()) {
                            callback(err(r.err));
                            error = just(r.err);
                        }
                        break;
                    }
                    case 'Ok': {
                        assign(r.value);
                        done();
                        break;
                    }
                }
            });
        }
        handle(this.t1, (x) => (ra = just(x)));
        handle(this.t2, (x) => (rb = just(x)));
    }
}
class TLambda extends Task {
    constructor(f) {
        super();
        this.f = f;
    }
    execute(callback) {
        try {
            callback(ok(this.f()));
        }
        catch (e) {
            callback(err(asError(e)));
        }
    }
}
class TPromise extends Task {
    constructor(p) {
        super();
        this.p = p;
    }
    execute(callback) {
        this.p()
            .then((r) => callback(ok(r)), (e) => callback(err(e)))
            .catch((e) => callback(err(e)));
    }
}
class TThen extends Task {
    constructor(task, f) {
        super();
        this.task = task;
        this.f = f;
    }
    execute(callback) {
        this.task.execute((r) => {
            r.match((r) => {
                const next = this.f(r);
                next.execute(callback);
            }, (e) => {
                callback(new Err(e));
            });
        });
    }
}
class TMapped extends Task {
    constructor(task, mapper) {
        super();
        this.task = task;
        this.mapper = mapper;
    }
    execute(callback) {
        this.task.execute((r) => {
            callback(r.map(this.mapper));
        });
    }
}
class TMappedErr extends Task {
    constructor(task, mapper) {
        super();
        this.task = task;
        this.mapper = mapper;
    }
    execute(callback) {
        this.task.execute((r) => {
            callback(r.mapError(this.mapper));
        });
    }
}
class TSuccess extends Task {
    constructor(result) {
        super();
        this.result = result;
    }
    execute(callback) {
        callback(ok(this.result()));
    }
}
class TError extends Task {
    constructor(err) {
        super();
        this.err = err;
    }
    execute(callback) {
        callback(err(this.err()));
    }
}
class TaskCmd extends Cmd {
    constructor(task, toMsg) {
        super();
        this.task = task;
        this.toMsg = toMsg;
    }
    execute(dispatch) {
        this.task.execute((r) => {
            dispatch(this.toMsg(r));
        });
    }
}
class TaskNoErrCmd extends Cmd {
    constructor(task, toMsg) {
        super();
        this.task = task;
        this.toMsg = toMsg;
    }
    execute(dispatch) {
        this.task.execute((r) => {
            r.match((ok) => dispatch(this.toMsg(ok)), (err) => {
                throw Error('got an error from a void task : ' + r + ', ' + err);
            });
        });
    }
}
