// tea-cup-core/dist/Time.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { Task } from './Task';
import { ok } from './Result';
import { Sub } from './Sub';
const everySubs = [];
const everyIntervals = [];
function initIntervalForDelay(delay) {
    if (everyIntervals[delay] === undefined) {
        everyIntervals[delay] = setInterval(() => {
            const subs = everySubs[delay];
            if (subs) {
                subs.forEach((sub) => sub.trigger());
            }
        }, delay);
    }
}
function cleanupIntervalForDelay(delay) {
    const handle = everyIntervals[delay];
    if (handle) {
        clearInterval(handle);
    }
}
class EverySub extends Sub {
    constructor(delay, toMsg) {
        super();
        this.delay = delay;
        this.toMsg = toMsg;
    }
    onInit() {
        super.onInit();
        let subsForDelay = everySubs[this.delay];
        if (subsForDelay === undefined) {
            subsForDelay = [];
            everySubs[this.delay] = subsForDelay;
        }
        subsForDelay.push(this);
        initIntervalForDelay(this.delay);
    }
    onRelease() {
        super.onRelease();
        let subsForDelay = everySubs[this.delay];
        if (subsForDelay !== undefined) {
            everySubs[this.delay] = subsForDelay.filter((s) => s !== this);
            if (everySubs[this.delay].length === 0) {
                cleanupIntervalForDelay(this.delay);
                delete everySubs[this.delay];
            }
        }
    }
    trigger() {
        this.dispatch(this.toMsg());
    }
}
/**
 * Simple module for getting current time and handling setTimeout
 */
export class Time {
    /**
     * Task that returns the current time
     */
    static now() {
        return new TimeTask();
    }
    /**
     * Task that fires in specified time
     */
    static in(timeout) {
        return new InTask(timeout);
    }
    static every(delay, toMsg) {
        return new EverySub(delay, toMsg);
    }
}
function now() {
    return new Date().getTime();
}
class TimeTask extends Task {
    execute(callback) {
        callback(ok(now()));
    }
}
class InTask extends Task {
    constructor(timeout) {
        super();
        this.t = timeout;
    }
    execute(callback) {
        setTimeout(() => callback(ok(now())), this.t);
    }
}
