// tea-pop-core/dist/Box.js
"use strict";
/*
 * MIT License
 *
 * Copyright (c) 2020 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.box = exports.Box = void 0;
const Pos_1 = require("./Pos");
const Dim_1 = require("./Dim");
class Box {
    constructor(p, d) {
        this.p = p;
        this.d = d;
    }
    static fromDomRect(rect) {
        const { x, y, height, width } = rect;
        return box(Pos_1.pos(x, y), Dim_1.dim(width, height));
    }
    get topLeft() {
        return this.p;
    }
    get bottomRight() {
        return Pos_1.pos(this.right, this.bottom);
    }
    get left() {
        return this.p.x;
    }
    get right() {
        return this.left + this.d.w;
    }
    get top() {
        return this.p.y;
    }
    get bottom() {
        return this.top + this.d.h;
    }
    isPointInside(pos) {
        return !(pos.x < this.left ||
            pos.x > this.right ||
            pos.y < this.top ||
            pos.y > this.bottom);
    }
    isBoxInside(box) {
        return this.isPointInside(box.topLeft) && this.isPointInside(box.bottomRight);
    }
}
exports.Box = Box;
function box(p, d) {
    return new Box(p, d);
}
exports.box = box;
