// react-tea-cup/dist/DevTools.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { noCmd, ObjectSerializer } from 'tea-cup-core';
const snapshotKey = 'teaCupSnapshot';
export class DevTools {
    constructor(objectSerializer) {
        this.events = [];
        this.pausedOnEvent = -1;
        this.listeners = [];
        this.maxEvents = -1;
        this.objectSerializer = objectSerializer;
    }
    static init(window, objectSerializer) {
        const dt = new DevTools(objectSerializer || ObjectSerializer.withTeaCupClasses());
        // @ts-ignore
        window['teaCupDevTools'] = dt;
        return dt;
    }
    isPaused() {
        return this.pausedOnEvent !== -1;
    }
    connected(program) {
        this.program = program;
    }
    onEvent(e) {
        this.events.push(e);
        this.removeEventsIfNeeded();
        this.listeners.forEach((l) => l(e));
    }
    travelTo(evtNum) {
        if (this.program) {
            const evt = this.events[evtNum];
            if (evt) {
                const model = this.getEventModel(evt);
                this.pausedOnEvent = evtNum;
                this.program.setModel(model, false);
            }
        }
    }
    getEventModel(e) {
        switch (e.tag) {
            case 'init':
                return e.model;
            case 'updated':
                return e.modelAfter;
        }
    }
    resume() {
        if (this.program) {
            if (this.events.length > 0) {
                const lastEvent = this.events[this.events.length - 1];
                if (lastEvent) {
                    const model = this.getEventModel(lastEvent);
                    this.pausedOnEvent = -1;
                    this.program.setModel(model, true);
                }
            }
        }
    }
    forward() {
        if (this.pausedOnEvent >= 0 && this.pausedOnEvent < this.events.length - 1) {
            this.travelTo(this.pausedOnEvent + 1);
        }
    }
    backward() {
        if (this.pausedOnEvent >= 1) {
            this.travelTo(this.pausedOnEvent - 1);
        }
    }
    addListener(l) {
        this.listeners.push(l);
    }
    removeListener(l) {
        this.listeners = this.listeners.filter((x) => x !== l);
    }
    lastEvent() {
        return this.events[this.events.length - 1];
    }
    lastModel() {
        const e = this.lastEvent();
        switch (e.tag) {
            case 'init': {
                return e.model;
            }
            case 'updated': {
                return e.modelAfter;
            }
        }
    }
    snapshot() {
        localStorage.setItem(snapshotKey, this.objectSerializer.serialize(this.lastModel()));
        console.log('********************************************************************\n' +
            '*** The current application state has been saved to local storage.\n' +
            '*** The application will now load with this initial state.\n' +
            "*** Call 'teaCupDevTools.clearSnapshot()' to restore normal loading.\n" +
            '********************************************************************');
    }
    initFromSnapshot() {
        const json = localStorage.getItem(snapshotKey);
        if (json) {
            try {
                // @ts-ignore
                const model = this.objectSerializer.deserialize(json);
                console.log('**********************************************************************************************\n' +
                    "*** The application has initialized from a state saved by calling 'teaCupDevTools.snapshot()'.\n" +
                    "*** Call 'teaCupDevTools.clearSnapshot()' if you want to restore normal loading.\n" +
                    '**********************************************************************************************');
                return noCmd(model);
            }
            catch (e) {
                console.log('*** Error restoring state from local storage: ', e);
            }
        }
    }
    clearSnapshot() {
        localStorage.removeItem(snapshotKey);
        console.log('***********************************************************************\n' +
            '*** Application state cleared, the application will now start normally.\n' +
            '***********************************************************************');
    }
    clear() {
        if (this.isPaused()) {
            this.resume();
        }
        this.events = [];
        console.log('All events cleared');
    }
    setMaxEvents(maxEvents) {
        this.maxEvents = maxEvents;
        this.removeEventsIfNeeded();
        return this;
    }
    removeEventsIfNeeded() {
        if (this.maxEvents > 0) {
            const delta = this.events.length - this.maxEvents;
            if (delta > 0) {
                const newEvents = this.events.slice(delta, this.events.length);
                this.events = newEvents;
            }
        }
        return this;
    }
}
