// react-tea-cup/dist/DocumentEvents.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { Sub } from 'tea-cup-core';
let subCount = 0;
class DocSub extends Sub {
    constructor(documentEvents, key, mapper, options) {
        super();
        this.documentEvents = documentEvents;
        this.key = key;
        this.mapper = mapper;
        this.options = options;
        subCount++;
        this.sc = subCount;
        this.listener = (e) => this.isActive() &&
            setTimeout(() => {
                this.dispatch(this.mapper(e));
            });
    }
    onInit() {
        super.onInit();
        this.documentEvents.doAddListener(this.key, this.listener, this.options);
    }
    onRelease() {
        super.onRelease();
        this.documentEvents.doRemoveListener(this.key, this.listener, this.options);
    }
}
class EventMapEvents {
    /**
     * Subscribe to an event.
     * @param key the event type to subscribe to.
     * @param mapper map the event to a message.
     * @param options options for this listener
     */
    on(key, mapper, options) {
        return new DocSub(this, key, mapper, options);
    }
}
/**
 * Subscribe to document events.
 */
export class DocumentEvents extends EventMapEvents {
    doAddListener(key, listener, options) {
        document.addEventListener(key, listener, options);
    }
    doRemoveListener(key, listener, options) {
        document.removeEventListener(key, listener, options);
    }
}
/**
 * Bonus, WindowEvents
 */
export class WindowEvents extends EventMapEvents {
    doAddListener(key, listener, options) {
        window.addEventListener(key, listener, options);
    }
    doRemoveListener(key, listener, options) {
        window.removeEventListener(key, listener, options);
    }
}
