// react-tea-cup/dist/Navigation.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
import { Program } from './Program';
import { List, Task, Ok, just, maybeOf, nothing } from 'tea-cup-core';
import * as React from 'react';
import { Component, createRef } from 'react';
/**
 * Program that handles navigation (routing).
 */
export class ProgramWithNav extends Component {
    constructor(props) {
        super(props);
        this.ref = createRef();
        this.listener = nothing;
    }
    render() {
        return (React.createElement(Program, { init: () => this.props.init(window.location), view: this.props.view, update: this.props.update, subscriptions: this.props.subscriptions, devTools: this.props.devTools, ref: this.ref }));
    }
    componentDidMount() {
        const l = () => {
            if (this.ref.current) {
                this.ref.current.dispatch(this.props.onUrlChange(window.location));
            }
        };
        this.listener = just(l);
        window.addEventListener('popstate', l);
    }
    componentWillUnmount() {
        if (this.listener.type === 'Just') {
            window.removeEventListener('popstate', this.listener.value);
            this.listener = nothing;
        }
    }
}
/**
 * Return a Task that will eventually change the browser location via historty.pushState,
 * and send back a Msg into the Program
 * @param url the url to navigate to
 */
export function newUrl(url) {
    return new NewUrlTask(url);
}
class NewUrlTask extends Task {
    constructor(url) {
        super();
        this.url = url;
    }
    execute(callback) {
        const state = {};
        window.history.pushState(state, '', this.url);
        const popStateEvent = new PopStateEvent('popstate', { state: state });
        dispatchEvent(popStateEvent);
        callback(new Ok(document.location));
    }
}
// router
// ------
export class PathElem {
}
class ConstantPathElem extends PathElem {
    constructor(s) {
        super();
        this.s = s;
    }
    mapPart(part) {
        if (part === this.s) {
            return just(part);
        }
        else {
            return nothing;
        }
    }
}
class RegexPathElem extends PathElem {
    constructor(regex, converter) {
        super();
        this.regex = regex;
        this.converter = converter;
    }
    mapPart(part) {
        if (this.regex.test(part)) {
            return this.converter(part);
        }
        else {
            return nothing;
        }
    }
}
class IntPathElem extends RegexPathElem {
    constructor() {
        super(new RegExp('^\\d+$'), (s) => {
            try {
                const i = parseInt(s, 10);
                if (isNaN(i)) {
                    return nothing;
                }
                else {
                    return just(i);
                }
            }
            catch (e) {
                return nothing;
            }
        });
    }
}
IntPathElem.INSTANCE = new IntPathElem();
class StrPathElem extends PathElem {
    mapPart(part) {
        return just(part);
    }
}
export function str(s) {
    if (s === undefined) {
        return new StrPathElem();
    }
    else {
        return new ConstantPathElem(s);
    }
}
export function int() {
    return IntPathElem.INSTANCE;
}
export function regex(r, converter) {
    return new RegexPathElem(r, converter);
}
export class Path0 {
    map(f) {
        return new RouteDef([], f);
    }
}
export class QueryParams {
    constructor(store, hash) {
        this.store = store;
        this.hash = hash;
    }
    getValues(name) {
        return maybeOf(this.store[name]);
    }
    getValue(name) {
        const values = this.store[name];
        if (values === undefined) {
            return nothing;
        }
        else {
            return List.fromArray(values).head();
        }
    }
    getHash() {
        return this.hash;
    }
    static empty() {
        return new QueryParams({}, nothing);
    }
    static fromQueryStringAndHash(queryString, hash) {
        const params = queryString === undefined ? [] : queryString.split('&');
        const store = {};
        function addToStore(name, value) {
            let values = store[name];
            if (values === undefined) {
                values = [value];
                store[name] = values;
            }
            else {
                values.push(value);
            }
        }
        params.forEach((param) => {
            const parts = param.split('=');
            if (parts.length === 1) {
                addToStore(parts[0], '');
            }
            else if (parts.length > 1) {
                addToStore(parts[0], parts[1]);
            }
        });
        return new QueryParams(store, hash === '' ? nothing : maybeOf(hash).map((h) => (h.startsWith('#') ? h.substring(1) : h)));
    }
}
export class Path1 {
    constructor(e) {
        this.e = e;
    }
    map(f) {
        return new RouteDef([this.e], f);
    }
}
export class Path2 {
    constructor(e1, e2) {
        this.e1 = e1;
        this.e2 = e2;
    }
    map(f) {
        return new RouteDef([this.e1, this.e2], f);
    }
}
export class Path3 {
    constructor(e1, e2, e3) {
        this.e1 = e1;
        this.e2 = e2;
        this.e3 = e3;
    }
    map(f) {
        return new RouteDef([this.e1, this.e2, this.e3], f);
    }
}
export const route0 = new Path0();
export function route1(e) {
    return new Path1(e);
}
export function route2(e1, e2) {
    return new Path2(e1, e2);
}
export function route3(e1, e2, e3) {
    return new Path3(e1, e2, e3);
}
export class RouteDef {
    constructor(pathElems, f) {
        this.pathElems = pathElems;
        this.f = f;
    }
    static sanitizePath(path) {
        const p1 = path.startsWith('/') ? path.substring(1) : path;
        return p1.endsWith('/') ? p1.substring(0, p1.length - 1) : p1;
    }
    static splitPath(path) {
        const p = RouteDef.sanitizePath(path);
        return p === '' ? [] : p.split('/').map(decodeURIComponent);
    }
    checkRoute(pathname, query) {
        // extract path parts from location and split
        const parts = RouteDef.splitPath(pathname);
        if (parts.length === this.pathElems.length) {
            // map every individual part, bail out if
            // something cannot be converted
            const mappedParts = [];
            for (let i = 0; i < parts.length; i++) {
                const part = parts[i];
                const pe = this.pathElems[i];
                const mapped = pe.mapPart(part);
                switch (mapped.type) {
                    case 'Just':
                        mappedParts.push(mapped.value);
                        break;
                    case 'Nothing':
                        return nothing;
                }
            }
            // append query params to args
            mappedParts.push(query);
            // now we have mapped args, let's call the route's func
            return just(this.f.apply({}, mappedParts));
        }
        else {
            return nothing;
        }
    }
}
export class Router {
    constructor(...routeDefs) {
        this.routes = routeDefs;
    }
    parse(pathname, query) {
        // try all routes one after the other
        for (let i = 0; i < this.routes.length; i++) {
            const d = this.routes[i];
            const r = d.checkRoute(pathname, query);
            if (r.type === 'Just') {
                return r;
            }
        }
        return nothing;
    }
    parseLocation(location) {
        return this.parse(location.pathname, QueryParams.fromQueryStringAndHash(location.search, location.hash));
    }
}
