// @diesel-parser/json-schema-facade-ts/dist/index.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
const json_schema_facade_js_1 = require("@diesel-parser/json-schema-facade-js");
function setLang(language) {
    // @ts-ignore
    return json_schema_facade_js_1.JsonSchemaJsFacade.setLang(language);
}
exports.setLang = setLang;
function validate(schema, value) {
    // @ts-ignore
    return json_schema_facade_js_1.JsonSchemaJsFacade.validate(schema, value);
}
exports.validate = validate;
function getErrors(res) {
    // @ts-ignore
    return json_schema_facade_js_1.JsonSchemaJsFacade.getErrors(res);
}
exports.getErrors = getErrors;
function propose(res, path, maxDepth = -1) {
    // @ts-ignore
    return json_schema_facade_js_1.JsonSchemaJsFacade.propose(res, path, maxDepth);
}
exports.propose = propose;
function getRenderers(res) {
    // @ts-ignore
    return json_schema_facade_js_1.JsonSchemaJsFacade.getRenderers(res);
}
exports.getRenderers = getRenderers;
function getFormats(res, path) {
    // @ts-ignore
    return json_schema_facade_js_1.JsonSchemaJsFacade.getFormats(res, path);
}
exports.getFormats = getFormats;
function getJsonParser(schema) {
    return json_schema_facade_js_1.JsonSchemaJsFacade.getJsonParser(schema);
}
exports.getJsonParser = getJsonParser;
