// @diesel-parser/json-form/dist/Actions.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setRoot = exports.actionUpdateValue = exports.actionMoveValue = exports.actionToggleExpandCollapsePath = exports.actionTriggerClicked = exports.updateMenu = exports.actionAddElementToArray = exports.actionAddProperty = exports.actionConfirmAddProperty = exports.actionAddPropertyClicked = exports.actionApplyProposal = exports.actionDeleteValue = void 0;
const Msg_1 = require("./Msg");
const tea_cup_core_1 = require("tea-cup-core");
const Model_1 = require("./Model");
const JsonValue_1 = require("./JsonValue");
const TPM = __importStar(require("tea-pop-menu"));
const ContextMenuActions_1 = require("./ContextMenuActions");
const Debouncer_1 = require("./Debouncer");
const JsFacade = __importStar(require("@diesel-parser/json-schema-facade-ts"));
function actionDeleteValue(model, path) {
    return setRoot(model, (0, JsonValue_1.deleteValueAt)(model.root.b, path).withDefault(model.root.b));
}
exports.actionDeleteValue = actionDeleteValue;
function actionApplyProposal(model, path, proposal, proposalIndex) {
    switch (proposal.tag) {
        case 'jv-object': {
            const newProposal = (0, JsonValue_1.getValueAt)(model.root.b, path)
                .map((valueAtPath) => {
                const augmentedProposal = model.validationResult
                    .andThen((vr) => {
                    // TODO deep propose only for proposalIndex
                    const all = JsFacade.propose(vr, path.format(), 5);
                    return (0, tea_cup_core_1.maybeOf)(all[proposalIndex]);
                })
                    .andThen((proposalAny) => {
                    return (0, JsonValue_1.valueFromAny)(proposalAny).match((jsonValue) => (0, tea_cup_core_1.just)(jsonValue), () => tea_cup_core_1.nothing);
                })
                    .andThen((v) => v.tag === 'jv-object' ? (0, tea_cup_core_1.just)(v) : tea_cup_core_1.nothing)
                    .withDefault(proposal);
                if (valueAtPath.tag === 'jv-object') {
                    // do not overwrite existing props
                    return (0, JsonValue_1.mergeProperties)(augmentedProposal, valueAtPath);
                }
                else {
                    return augmentedProposal;
                }
            })
                .withDefault(proposal);
            return doUpdateValue(model, path, newProposal);
        }
        default: {
            return doUpdateValue(model, path, proposal);
        }
    }
}
exports.actionApplyProposal = actionApplyProposal;
function doUpdateValue(model, path, value) {
    if (path.isEmpty()) {
        return setRoot(model, value);
    }
    return setRoot(model, (0, JsonValue_1.setValueAt)(model.root.b, path, value));
}
function actionAddPropertyClicked(model, ownerPath) {
    const focusTask = tea_cup_core_1.Task.succeedLazy(() => {
        const input = document.getElementById('property-name-editor');
        if (input) {
            input.focus();
        }
    });
    return tea_cup_core_1.Tuple.t2n(Object.assign(Object.assign({}, model), { adding: (0, tea_cup_core_1.just)({
            ownerPath,
            addingPropName: '',
            isDuplicate: false,
        }) }), tea_cup_core_1.Task.perform(focusTask, () => Msg_1.noOp));
}
exports.actionAddPropertyClicked = actionAddPropertyClicked;
function actionConfirmAddProperty(model, commit) {
    const newModel = Object.assign(Object.assign({}, model), { adding: tea_cup_core_1.nothing });
    if (!commit) {
        return (0, tea_cup_core_1.noCmd)(newModel);
    }
    return model.adding
        .filter((addingState) => addingState.addingPropName !== '')
        .andThen((addingState) => {
        return (0, JsonValue_1.mapValueAt)(model.root.b, addingState.ownerPath, (owningValue) => {
            if (owningValue.tag === 'jv-object') {
                return (0, tea_cup_core_1.just)((0, JsonValue_1.jvObject)([
                    ...owningValue.properties,
                    { name: addingState.addingPropName, value: JsonValue_1.jvNull },
                ]));
            }
            return (0, tea_cup_core_1.just)(owningValue);
        }).map((newRoot) => setRoot(newModel, newRoot));
    })
        .withDefaultSupply(() => (0, tea_cup_core_1.noCmd)(newModel));
}
exports.actionConfirmAddProperty = actionConfirmAddProperty;
function actionAddProperty(model, path, propertyName) {
    return (0, JsonValue_1.getValueAt)(model.root.b, path)
        .map((owner) => {
        if (owner.tag === 'jv-object') {
            // create the new object with a null value
            // because we need it to propose
            const newObject = (0, JsonValue_1.jvObject)([
                ...owner.properties,
                { name: propertyName, value: JsonValue_1.jvNull },
            ]);
            const newRoot = (0, JsonValue_1.setValueAt)(model.root.b, path, newObject);
            const newValidationResult = model.schema
                .map((s) => s.a)
                .map((schemaAny) => JsFacade.validate(schemaAny, (0, JsonValue_1.valueToAny)(newRoot)));
            const propertyProposals = newValidationResult
                .map((vr) => (0, Model_1.getProposals)(vr, path.append(propertyName), -1))
                .withDefault([])
                .map(JsonValue_1.clearPropertiesIfObject);
            const newObject2 = (0, JsonValue_1.jvObject)([
                ...owner.properties,
                {
                    name: propertyName,
                    value: propertyProposals.length === 0 ? JsonValue_1.jvNull : propertyProposals[0],
                },
            ]);
            return setRoot(model, (0, JsonValue_1.setValueAt)(model.root.b, path, newObject2));
        }
        return (0, tea_cup_core_1.noCmd)(model);
    })
        .withDefaultSupply(() => (0, tea_cup_core_1.noCmd)(model));
}
exports.actionAddProperty = actionAddProperty;
function actionAddElementToArray(model, path) {
    return (0, JsonValue_1.getValueAt)(model.root.b, path)
        .map((array) => {
        if (array.tag === 'jv-array') {
            const newElemIndex = array.elems.length;
            // we create a transient JsonValue with the array updated
            // so that we have a value at new index path
            // otherwise the proposals would be empty because
            // no path matches the requested index
            const tmpArray = (0, JsonValue_1.jvArray)([...array.elems, JsonValue_1.jvNull]);
            const tmpRoot = (0, JsonValue_1.setValueAt)(model.root.b, path, tmpArray);
            const newValidationResult = model.schema
                .map((s) => s.a)
                .map((schemaAny) => JsFacade.validate(schemaAny, (0, JsonValue_1.valueToAny)(tmpRoot)));
            const proposals = newValidationResult
                .map((vr) => (0, Model_1.getProposals)(vr, path.append(newElemIndex), -1))
                .withDefault([]);
            const proposal = (0, tea_cup_core_1.maybeOf)(proposals[0]).withDefault(JsonValue_1.jvNull);
            const newArray = Object.assign(Object.assign({}, array), { elems: [...array.elems, (0, JsonValue_1.clearPropertiesIfObject)(proposal)] });
            const newRoot = (0, JsonValue_1.setValueAt)(model.root.b, path, newArray);
            return setRoot(model, newRoot);
        }
        return (0, tea_cup_core_1.noCmd)(model);
    })
        .withDefaultSupply(() => (0, tea_cup_core_1.noCmd)(model));
}
exports.actionAddElementToArray = actionAddElementToArray;
function updateMenu(model, mac) {
    return tea_cup_core_1.Tuple.fromNative(mac)
        .mapFirst((mm) => {
        return Object.assign(Object.assign({}, model), { menuModel: (0, tea_cup_core_1.just)(mm) });
    })
        .mapSecond((mc) => mc.map(Msg_1.contextMenuMsg))
        .toNative();
}
exports.updateMenu = updateMenu;
function actionTriggerClicked(model, path, refBox, menuFilter) {
    return (0, JsonValue_1.getValueAt)(model.root.b, path)
        .map((valueAtPath) => {
        const focusMenuCmd = tea_cup_core_1.Task.attempt(tea_cup_core_1.Task.fromLambda(() => {
            const e = document.getElementById('dummy-textarea');
            if (e) {
                e.focus();
            }
        }), () => {
            return Msg_1.noOp;
        });
        const mac = updateMenu(model, TPM.open((0, ContextMenuActions_1.createMenu)({
            root: model.root.b,
            path,
            proposals: model.validationResult
                .map((vr) => (0, Model_1.getProposals)(vr, path, -1))
                .withDefault([]),
            valueAtPath,
            strictMode: model.strictMode,
            menuFilter,
        }), refBox));
        return tea_cup_core_1.Tuple.t2n(mac[0], tea_cup_core_1.Cmd.batch([mac[1], focusMenuCmd]));
    })
        .withDefaultSupply(() => (0, tea_cup_core_1.noCmd)(model));
}
exports.actionTriggerClicked = actionTriggerClicked;
function actionToggleExpandCollapsePath(model, path) {
    const pathStr = path.format();
    const collapsedPaths = model.collapsedPaths.has(pathStr)
        ? new Set(Array.from(model.collapsedPaths.keys()).filter((s) => s !== pathStr))
        : new Set(Array.from(model.collapsedPaths.keys()).concat(pathStr));
    return (0, tea_cup_core_1.noCmd)(Object.assign(Object.assign({}, model), { collapsedPaths }));
}
exports.actionToggleExpandCollapsePath = actionToggleExpandCollapsePath;
function actionMoveValue(model, path, direction) {
    return path
        .parent()
        .andThen((parentPath) => (0, JsonValue_1.getValueAt)(model.root.b, parentPath).andThen((parentValue) => path.lastElem().map((lastPathElem) => {
        switch (parentValue.tag) {
            case 'jv-object': {
                const newRoot = (0, JsonValue_1.moveProperty)(model.root.b, parentPath, lastPathElem, direction);
                return setRoot(model, newRoot);
            }
            case 'jv-array': {
                const index = parseInt(lastPathElem);
                if (isNaN(index)) {
                    return (0, tea_cup_core_1.noCmd)(model);
                }
                const newRoot = (0, JsonValue_1.moveElement)(model.root.b, parentPath, index, direction);
                return setRoot(model, newRoot);
            }
            default: {
                return (0, tea_cup_core_1.noCmd)(model);
            }
        }
    })))
        .withDefaultSupply(() => (0, tea_cup_core_1.noCmd)(model));
}
exports.actionMoveValue = actionMoveValue;
function actionUpdateValue(model, path, value) {
    return doUpdateValue(model, path, value);
}
exports.actionUpdateValue = actionUpdateValue;
const debouncer = new Debouncer_1.Debouncer();
function setRoot(model, root) {
    const cmd = debouncer.debounce({ tag: 'recompute-metadata' }, model.debounceMs);
    const newModel = Object.assign(Object.assign({}, model), { root: new tea_cup_core_1.Tuple((0, JsonValue_1.valueToAny)(root), root), adding: tea_cup_core_1.nothing });
    return [newModel, cmd];
}
exports.setRoot = setRoot;
