// @diesel-parser/json-form/dist/ContextMenuRenderer.test.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const ContextMenuActions_1 = require("./ContextMenuActions");
const JsonValue_1 = require("./JsonValue");
const JsPath_1 = require("./JsPath");
const tea_pop_menu_1 = require("tea-pop-menu");
const RenderOptions_1 = require("./RenderOptions");
describe('Change type menu', () => {
    test('[SM: ON, P: 1, V: valid] ' + 'no types menu', () => {
        const typesMenu = buildTypesMenuStrictMode([(0, JsonValue_1.jvNumber)('123'), (0, JsonValue_1.jvNumber)('456')], (0, JsonValue_1.jvNumber)('456'));
        expect(typesMenu).toStrictEqual([]);
    });
    test('[SM: ON, P: 2 same type, V: valid] ' + 'no types menu', () => {
        const typesMenu = buildTypesMenuStrictMode([(0, JsonValue_1.jvNumber)('123'), (0, JsonValue_1.jvNumber)('456')], (0, JsonValue_1.jvNumber)('456'));
        expect(typesMenu).toStrictEqual([]);
    });
    test('[SM: ON, P: 2 different type, V: something different than proposal] ' +
        'menu type exists and subMenu should contains 2 entries corresponding to the 2 proposals', () => {
        const typesMenu = buildTypesMenuStrictMode([(0, JsonValue_1.jvNumber)('123'), JsonValue_1.jvNull], (0, JsonValue_1.jvString)('Foo'));
        const expectedTypesMenu = buildExpectedTypesMenu([(0, JsonValue_1.jvNumber)('0'), JsonValue_1.jvNull]);
        expect(typesMenu).toStrictEqual(expectedTypesMenu);
    });
    test('[SM: ON, P: 2 different type, V: something equals than proposal] ' +
        'menu type exists and subMenu should contains 1 entry corresponding to the proposal that does not match the value', () => {
        const typesMenu = buildTypesMenuStrictMode([(0, JsonValue_1.jvNumber)('123'), JsonValue_1.jvNull], (0, JsonValue_1.jvNumber)('456'));
        const expectedTypesMenu = buildExpectedTypesMenu([JsonValue_1.jvNull]);
        expect(typesMenu).toStrictEqual(expectedTypesMenu);
    });
    test('[SM: ON, P: 1, V: wrong type]' +
        'menu type exists and subMenu should contains 1 entry corresponding to the proposal', () => {
        const typesMenu = buildTypesMenuStrictMode([(0, JsonValue_1.jvNumber)('123')], JsonValue_1.jvNull);
        const expectedTypesMenu = buildExpectedTypesMenu([(0, JsonValue_1.jvNumber)('0')]);
        expect(typesMenu).toStrictEqual(expectedTypesMenu);
    });
    test('[SM: OFF, whatever] ' +
        'menu type exists and subMenu should contains all available types', () => {
        const typesMenu = buildTypesMenu(JsonValue_1.jvNull);
        const expectedTypesMenu = buildExpectedTypesMenu([
            (0, JsonValue_1.jvString)(''),
            (0, JsonValue_1.jvNumber)('0'),
            (0, JsonValue_1.jvBool)(true),
            (0, JsonValue_1.jvObject)(),
            (0, JsonValue_1.jvArray)(),
        ]);
        expect(typesMenu).toStrictEqual(expectedTypesMenu);
    });
});
describe('menu render options', () => {
    describe('default options', () => {
        test('null value', () => {
            const props = {
                root: JsonValue_1.jvNull,
                path: JsPath_1.JsPath.empty,
                valueAtPath: JsonValue_1.jvNull,
                proposals: [],
                strictMode: false,
                menuFilter: {},
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(menuActionTags(actual)).toEqual(['types', 'delete']);
        });
        test('empty object value', () => {
            const props = {
                root: (0, JsonValue_1.jvObject)(),
                path: JsPath_1.JsPath.empty,
                valueAtPath: (0, JsonValue_1.jvObject)(),
                proposals: [],
                strictMode: false,
                menuFilter: {},
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(menuActionTags(actual)).toEqual(['add', 'types', 'delete']);
        });
        test('empty array value', () => {
            const props = {
                root: (0, JsonValue_1.jvArray)(),
                path: JsPath_1.JsPath.empty,
                valueAtPath: (0, JsonValue_1.jvArray)(),
                proposals: [],
                strictMode: false,
                menuFilter: {},
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(menuActionTags(actual)).toEqual(['add', 'types', 'delete']);
        });
        test('value with proposals', () => {
            const props = {
                root: (0, JsonValue_1.jvArray)(),
                path: JsPath_1.JsPath.empty,
                valueAtPath: JsonValue_1.jvNull,
                proposals: [JsonValue_1.jvNull],
                strictMode: false,
                menuFilter: {},
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(menuActionTags(actual)).toEqual(['types', 'propose', 'delete']);
        });
        test('non empty object value with move', () => {
            const props = {
                root: (0, JsonValue_1.jvObject)([
                    { name: 'foo', value: JsonValue_1.jvNull },
                    { name: 'bar', value: JsonValue_1.jvNull },
                ]),
                path: JsPath_1.JsPath.empty.append('bar'),
                valueAtPath: JsonValue_1.jvNull,
                proposals: [],
                strictMode: false,
                menuFilter: {},
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(menuActionTags(actual)).toEqual(['move', 'types', 'delete']);
        });
        test('non empty array value with move', () => {
            const props = {
                root: (0, JsonValue_1.jvArray)([JsonValue_1.jvNull, JsonValue_1.jvNull]),
                path: JsPath_1.JsPath.empty.append(1),
                valueAtPath: JsonValue_1.jvNull,
                proposals: [],
                strictMode: false,
                menuFilter: {},
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(menuActionTags(actual)).toEqual(['move', 'types', 'delete']);
        });
    });
    describe('menu filtering options', () => {
        test('null value no delete', () => {
            const props = {
                root: JsonValue_1.jvNull,
                path: JsPath_1.JsPath.empty,
                valueAtPath: JsonValue_1.jvNull,
                proposals: [],
                strictMode: false,
                menuFilter: {
                    menuFilters: RenderOptions_1.MenuOptions.DELETE,
                },
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.ADD)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.CHANGE_TYPE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.DELETE)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.MOVE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.PROPOSE)).toEqual(false);
            expect(menuActionTags(actual)).toEqual(['types']);
        });
        test('empty object value no add or change type', () => {
            const props = {
                root: (0, JsonValue_1.jvObject)(),
                path: JsPath_1.JsPath.empty,
                valueAtPath: (0, JsonValue_1.jvObject)(),
                proposals: [],
                strictMode: false,
                menuFilter: {
                    menuFilters: RenderOptions_1.MenuOptions.ADD,
                },
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.ADD)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.CHANGE_TYPE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.DELETE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.MOVE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.PROPOSE)).toEqual(false);
            expect(menuActionTags(actual)).toEqual(['types', 'delete']);
        });
        test('empty array value no delete or change type', () => {
            const props = {
                root: (0, JsonValue_1.jvArray)(),
                path: JsPath_1.JsPath.empty,
                valueAtPath: (0, JsonValue_1.jvArray)(),
                proposals: [],
                strictMode: false,
                menuFilter: {
                    menuFilters: RenderOptions_1.MenuOptions.DELETE | RenderOptions_1.MenuOptions.CHANGE_TYPE,
                },
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.ADD)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.CHANGE_TYPE)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.DELETE)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.MOVE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.PROPOSE)).toEqual(false);
            expect(menuActionTags(actual)).toEqual(['add']);
        });
        test('value with proposals no propose or change type or add', () => {
            const props = {
                root: (0, JsonValue_1.jvArray)(),
                path: JsPath_1.JsPath.empty,
                valueAtPath: JsonValue_1.jvNull,
                proposals: [JsonValue_1.jvNull],
                strictMode: false,
                menuFilter: {
                    menuFilters: RenderOptions_1.MenuOptions.CHANGE_TYPE | RenderOptions_1.MenuOptions.PROPOSE | RenderOptions_1.MenuOptions.ADD,
                },
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.ADD)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.CHANGE_TYPE)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.DELETE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.MOVE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.PROPOSE)).toEqual(true);
            expect(menuActionTags(actual)).toEqual(['delete']);
        });
        test('non empty object value with move no change type', () => {
            const props = {
                root: (0, JsonValue_1.jvObject)([
                    { name: 'foo', value: JsonValue_1.jvNull },
                    { name: 'bar', value: JsonValue_1.jvNull },
                ]),
                path: JsPath_1.JsPath.empty.append('bar'),
                valueAtPath: JsonValue_1.jvNull,
                proposals: [],
                strictMode: false,
                menuFilter: {
                    menuFilters: RenderOptions_1.MenuOptions.CHANGE_TYPE,
                },
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.ADD)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.CHANGE_TYPE)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.DELETE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.MOVE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.PROPOSE)).toEqual(false);
            expect(menuActionTags(actual)).toEqual(['move', 'delete']);
        });
        test('non empty array value with move no change type or move', () => {
            const props = {
                root: (0, JsonValue_1.jvArray)([JsonValue_1.jvNull, JsonValue_1.jvNull]),
                path: JsPath_1.JsPath.empty.append(1),
                valueAtPath: JsonValue_1.jvNull,
                proposals: [],
                strictMode: false,
                menuFilter: {
                    menuFilters: RenderOptions_1.MenuOptions.MOVE | RenderOptions_1.MenuOptions.CHANGE_TYPE,
                },
            };
            const actual = (0, ContextMenuActions_1.createMenu)(props);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.ADD)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.CHANGE_TYPE)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.DELETE)).toEqual(false);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.MOVE)).toEqual(true);
            expect(getSelectedFilter(props, RenderOptions_1.MenuOptions.PROPOSE)).toEqual(false);
            expect(menuActionTags(actual)).toEqual(['delete']);
        });
    });
});
/*-----------*/
/*  Helpers  */
/*-----------*/
function buildTypesMenu(valueAtPath) {
    return (0, ContextMenuActions_1.createTypesMenu)(JsPath_1.JsPath.empty, valueAtPath, [], false);
}
function buildTypesMenuStrictMode(proposals, valueAtPath) {
    return (0, ContextMenuActions_1.createTypesMenu)(JsPath_1.JsPath.empty, valueAtPath, proposals, true);
}
function buildExpectedTypesMenu(values) {
    return [
        (0, tea_pop_menu_1.item)({ tag: 'types' }, (0, tea_pop_menu_1.menu)(values.map((v) => {
            return (0, tea_pop_menu_1.item)({ tag: 'change-type', path: JsPath_1.JsPath.empty, value: v });
        }))),
    ];
}
function menuActionTags(menu) {
    return menu.elems.flatMap((e) => {
        switch (e.tag) {
            case 'item':
                return [e.userData.tag];
            default:
                return [];
        }
    });
}
function getSelectedFilter(menuPropertyProps, filter) {
    var _a, _b;
    return (((_a = menuPropertyProps === null || menuPropertyProps === void 0 ? void 0 : menuPropertyProps.menuFilter) === null || _a === void 0 ? void 0 : _a.menuFilters) &&
        (((_b = menuPropertyProps === null || menuPropertyProps === void 0 ? void 0 : menuPropertyProps.menuFilter) === null || _b === void 0 ? void 0 : _b.menuFilters) & filter) === filter);
}
