// @diesel-parser/json-form/dist/ContextMenuActions.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMenu = exports.createProposeMenu = exports.createTypesMenu = void 0;
const tea_pop_menu_1 = require("tea-pop-menu");
const JsonValue_1 = require("./JsonValue");
const RenderOptions_1 = require("./RenderOptions");
function maDelete(path) {
    return {
        tag: 'delete',
        path,
    };
}
function maMove() {
    return {
        tag: 'move',
    };
}
function maMoveUp(path) {
    return {
        tag: 'move-up',
        path,
    };
}
function maMoveDown(path) {
    return {
        tag: 'move-down',
        path,
    };
}
function createTypesMenu(path, valueAtPath, proposals, strictMode) {
    const buildChangeTypeItem = (value) => {
        return (0, tea_pop_menu_1.item)({
            tag: 'change-type',
            path,
            value: value,
        });
    };
    const uniqueProposalTypes = proposals
        .map((p) => p.tag)
        .filter((value, index, array) => array.indexOf(value) === index);
    const buildChangeTypeMenuItems = () => {
        const jsonValues = strictMode
            ? proposals
            : [
                JsonValue_1.jvNull,
                (0, JsonValue_1.jvString)(''),
                (0, JsonValue_1.jvNumber)('0'),
                (0, JsonValue_1.jvBool)(true),
                (0, JsonValue_1.jvObject)(),
                (0, JsonValue_1.jvArray)(),
            ];
        return (jsonValues
            // keep only JsonValue type
            .map((p) => p.tag)
            // Get unique values
            .filter((value, index, array) => array.indexOf(value) === index)
            // Exclude actual type
            .filter((value) => value !== valueAtPath.tag)
            // Build MenuItem from unique JsonValue
            .map((t) => {
            switch (t) {
                case 'jv-array':
                    return buildChangeTypeItem((0, JsonValue_1.jvArray)());
                case 'jv-boolean':
                    return buildChangeTypeItem((0, JsonValue_1.jvBool)(true));
                case 'jv-null':
                    return buildChangeTypeItem(JsonValue_1.jvNull);
                case 'jv-number':
                    return buildChangeTypeItem((0, JsonValue_1.jvNumber)('0'));
                case 'jv-object':
                    return buildChangeTypeItem((0, JsonValue_1.jvObject)());
                case 'jv-string':
                    return buildChangeTypeItem((0, JsonValue_1.jvString)(''));
            }
        }));
    };
    if (strictMode &&
        uniqueProposalTypes.length === 1 &&
        valueAtPath.tag === proposals[0].tag) {
        // Only one type accepted and the value has already the right one -> no menu at all
        return [];
    }
    return buildChangeTypeMenuItems().length > 0
        ? [(0, tea_pop_menu_1.item)({ tag: 'types' }, (0, tea_pop_menu_1.menu)(buildChangeTypeMenuItems()))]
        : [];
}
exports.createTypesMenu = createTypesMenu;
function createProposeMenu(path, proposals, strictMode) {
    if (strictMode || proposals.length === 0) {
        // Strict mode or no proposal => no menu
        return [];
    }
    const proposeMenu = (0, tea_pop_menu_1.menu)(proposals.map((value, index) => (0, tea_pop_menu_1.item)({
        tag: 'proposal',
        path,
        value,
        index,
    })));
    return [(0, tea_pop_menu_1.item)({ tag: 'propose', path }, proposeMenu)];
}
exports.createProposeMenu = createProposeMenu;
function createMenu(props) {
    const { path, proposals, valueAtPath, root, strictMode, menuFilter } = props;
    const addItems = () => {
        const isArray = valueAtPath.tag === 'jv-array';
        const isObject = !strictMode && valueAtPath.tag === 'jv-object';
        if ((isArray || isObject) &&
            !(0, RenderOptions_1.isMenuOptionHidden)(RenderOptions_1.MenuOptions.ADD, menuFilter)) {
            return [(0, tea_pop_menu_1.item)({ tag: 'add', path, isArray })];
        }
        return [];
    };
    const isRoot = path.isEmpty();
    const nbItems = path
        .parent()
        .andThen((pp) => (0, JsonValue_1.getValueAt)(root, pp))
        .map((pv) => {
        switch (pv.tag) {
            case 'jv-object':
                return pv.properties.length;
            case 'jv-array':
                return pv.elems.length;
            default:
                return 0;
        }
    })
        .withDefault(0);
    const hasMoveMenu = !isRoot && nbItems > 1 && !(0, RenderOptions_1.isMenuOptionHidden)(RenderOptions_1.MenuOptions.MOVE, menuFilter);
    const moveMenu = () => {
        const index = (0, JsonValue_1.indexOfPathInParent)(root, path);
        const canMoveUp = index > 0;
        const canMoveDown = index >= 0 && index < nbItems - 1;
        const moveUpItems = canMoveUp ? [(0, tea_pop_menu_1.item)(maMoveUp(path))] : [];
        const moveDownItems = canMoveDown ? [(0, tea_pop_menu_1.item)(maMoveDown(path))] : [];
        return (0, tea_pop_menu_1.menu)(moveUpItems.concat(moveDownItems));
    };
    const moveItems = hasMoveMenu ? [(0, tea_pop_menu_1.item)(maMove(), moveMenu())] : [];
    const deleteItems = isRoot &&
        (menuFilter === null || menuFilter === void 0 ? void 0 : menuFilter.menuFilters) &&
        (0, RenderOptions_1.isMenuOptionHidden)(RenderOptions_1.MenuOptions.DELETE, menuFilter)
        ? []
        : [(0, tea_pop_menu_1.item)(maDelete(path))];
    const changeTypes = (0, RenderOptions_1.isMenuOptionHidden)(RenderOptions_1.MenuOptions.CHANGE_TYPE, menuFilter)
        ? []
        : createTypesMenu(path, valueAtPath, proposals, strictMode);
    const proposeItems = (0, RenderOptions_1.isMenuOptionHidden)(RenderOptions_1.MenuOptions.PROPOSE, menuFilter)
        ? []
        : createProposeMenu(path, proposals, strictMode);
    return (0, tea_pop_menu_1.menu)(moveItems
        .concat(addItems())
        .concat(changeTypes)
        .concat(proposeItems)
        .concat(deleteItems));
}
exports.createMenu = createMenu;
