// @diesel-parser/json-form/dist/Debouncer.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Debouncer = void 0;
const tea_cup_core_1 = require("tea-cup-core");
class Debouncer {
    debounce(msg, delay) {
        if (this.cmd) {
            this.cmd.cancel();
            delete this.cmd;
        }
        this.cmd = new DebounceCmd(msg, delay);
        return this.cmd;
    }
}
exports.Debouncer = Debouncer;
class DebounceCmd extends tea_cup_core_1.Cmd {
    constructor(msg, delay) {
        super();
        this.msg = msg;
        this.delay = delay;
        this.canceled = false;
    }
    execute(dispatch) {
        this.timeout = setTimeout(() => {
            if (!this.canceled) {
                dispatch(this.msg);
            }
        }, this.delay);
    }
    cancel() {
        if (this.timeout) {
            clearTimeout(this.timeout);
        }
        this.canceled = true;
    }
}
