// @diesel-parser/json-form/dist/JsonValue.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearPropertiesIfObject = exports.indexOfPathInParent = exports.mergeProperties = exports.moveElement = exports.moveProperty = exports.mapValueAt = exports.getValueAt = exports.deleteValueAt = exports.setValueAt = exports.parseJsonValue = exports.valueToAny = exports.valueFromAny = exports.valueType = exports.jvArray = exports.jvObject = exports.jvBool = exports.jvNumber = exports.jvString = exports.jvNull = void 0;
const tea_cup_core_1 = require("tea-cup-core");
exports.jvNull = { tag: 'jv-null' };
function jvString(value) {
    return {
        tag: 'jv-string',
        value,
    };
}
exports.jvString = jvString;
function jvNumber(value) {
    return {
        tag: 'jv-number',
        value,
    };
}
exports.jvNumber = jvNumber;
function jvBool(value) {
    return {
        tag: 'jv-boolean',
        value,
    };
}
exports.jvBool = jvBool;
function jvObject(properties = []) {
    return {
        tag: 'jv-object',
        properties,
    };
}
exports.jvObject = jvObject;
function jvArray(elems = []) {
    return {
        tag: 'jv-array',
        elems,
    };
}
exports.jvArray = jvArray;
function valueType(value) {
    switch (value.tag) {
        case 'jv-string':
            return 'string';
        case 'jv-object':
            return 'object';
        case 'jv-number':
            return 'number';
        case 'jv-null':
            return 'null';
        case 'jv-boolean':
            return 'boolean';
        case 'jv-array':
            return 'array';
    }
}
exports.valueType = valueType;
function valueFromAny(value) {
    if (value === undefined) {
        return (0, tea_cup_core_1.err)('undefined');
    }
    if (value === null) {
        return (0, tea_cup_core_1.ok)(exports.jvNull);
    }
    if (Array.isArray(value)) {
        const a = value;
        const elems = [];
        for (let i = 0; i < a.length; i++) {
            const elemValue = valueFromAny(a[i]);
            switch (elemValue.tag) {
                case 'Ok': {
                    elems.push(elemValue.value);
                    break;
                }
                case 'Err': {
                    return elemValue;
                }
            }
        }
        return (0, tea_cup_core_1.ok)({
            tag: 'jv-array',
            elems,
        });
    }
    const valueType = typeof value;
    switch (valueType) {
        case 'object':
            if (!isObjLiteral(value)) {
                return (0, tea_cup_core_1.err)('not an object literal');
            }
            const keys = Object.keys(value);
            const properties = [];
            for (let i = 0; i < keys.length; i++) {
                const propName = keys[i];
                const propValue = valueFromAny(value[propName]);
                switch (propValue.tag) {
                    case 'Ok': {
                        properties.push({
                            name: propName,
                            value: propValue.value,
                        });
                        break;
                    }
                    case 'Err': {
                        return propValue;
                    }
                }
            }
            return (0, tea_cup_core_1.ok)({ tag: 'jv-object', properties });
        case 'string':
            return (0, tea_cup_core_1.ok)({ tag: 'jv-string', value });
        case 'boolean':
            return (0, tea_cup_core_1.ok)({ tag: 'jv-boolean', value });
        case 'number':
            return (0, tea_cup_core_1.ok)({ tag: 'jv-number', value: value.toString() });
        default:
            return (0, tea_cup_core_1.err)(valueType);
    }
}
exports.valueFromAny = valueFromAny;
function valueToAny(value) {
    switch (value.tag) {
        case 'jv-array':
            return value.elems.map(valueToAny);
        case 'jv-object': {
            const res = {};
            value.properties.forEach((p) => {
                res[p.name] = valueToAny(p.value);
            });
            return res;
        }
        case 'jv-boolean':
            return value.value;
        case 'jv-null':
            return null;
        case 'jv-number':
            const parsed = parseFloat(value.value);
            return isNaN(parsed) ? value.value : parsed;
        case 'jv-string': {
            return value.value;
        }
    }
}
exports.valueToAny = valueToAny;
function isObjLiteral(_obj) {
    let _test = _obj;
    return typeof _obj !== 'object' || _obj === null
        ? false
        : (function () {
            while (true) {
                if (Object.getPrototypeOf((_test = Object.getPrototypeOf(_test))) ===
                    null) {
                    break;
                }
            }
            return Object.getPrototypeOf(_obj) === _test;
        })();
}
function parseJsonValue(json) {
    try {
        const x = JSON.parse(json);
        return valueFromAny(x);
    }
    catch (e) {
        if (e instanceof Error) {
            return (0, tea_cup_core_1.err)(e.message);
        }
        return (0, tea_cup_core_1.err)('unhandled error ' + e);
    }
}
exports.parseJsonValue = parseJsonValue;
function setValueAt(root, path, value) {
    return mapValueAt(root, path, () => (0, tea_cup_core_1.just)(value)).withDefault(root);
}
exports.setValueAt = setValueAt;
function deleteValueAt(root, path) {
    return mapValueAt(root, path, () => tea_cup_core_1.nothing);
}
exports.deleteValueAt = deleteValueAt;
function getValueAt(root, path) {
    const ph = path.head();
    switch (ph.type) {
        case 'Just': {
            const pathHead = ph.value;
            switch (root.tag) {
                case 'jv-object':
                    const prop = root.properties.find((p) => p.name === pathHead);
                    if (prop === undefined) {
                        return tea_cup_core_1.nothing;
                    }
                    return getValueAt(prop.value, path.tail());
                case 'jv-array':
                    const index = parseInt(pathHead);
                    if (isNaN(index)) {
                        return tea_cup_core_1.nothing;
                    }
                    const elem = root.elems[index];
                    if (elem === undefined) {
                        return tea_cup_core_1.nothing;
                    }
                    return getValueAt(elem, path.tail());
                default:
                    return (0, tea_cup_core_1.just)(root);
            }
        }
        case 'Nothing': {
            // empty path
            return (0, tea_cup_core_1.just)(root);
        }
    }
}
exports.getValueAt = getValueAt;
function mapValueAt(root, path, f) {
    const ph = path.head();
    switch (ph.type) {
        case 'Nothing': {
            return f(root);
        }
        case 'Just': {
            const pathHead = ph.value;
            switch (root.tag) {
                case 'jv-object': {
                    return (0, tea_cup_core_1.just)(jvObject(root.properties.flatMap((prop) => {
                        if (prop.name === pathHead) {
                            return mapValueAt(prop.value, path.tail(), f)
                                .map((value) => [
                                Object.assign(Object.assign({}, prop), { value }),
                            ])
                                .withDefault([]);
                        }
                        else {
                            return [prop];
                        }
                    })));
                }
                case 'jv-array': {
                    const pathIndex = parseInt(pathHead);
                    if (isNaN(pathIndex)) {
                        return (0, tea_cup_core_1.just)(root);
                    }
                    return (0, tea_cup_core_1.just)(jvArray(root.elems.flatMap((elem, elemIndex) => {
                        if (elemIndex === pathIndex) {
                            return mapValueAt(elem, path.tail(), f)
                                .map((value) => [value])
                                .withDefault([]);
                        }
                        else {
                            return [elem];
                        }
                    })));
                }
                default: {
                    return (0, tea_cup_core_1.just)(root);
                }
                // }
            }
        }
    }
}
exports.mapValueAt = mapValueAt;
function moveElems(array, index, direction) {
    const newIndex = index + (direction === 'up' ? -1 : 1);
    if (newIndex < 0 || newIndex > array.length - 1) {
        return array;
    }
    const newArray = new Array(...array);
    newArray[index] = array[newIndex];
    newArray[newIndex] = array[index];
    return newArray;
}
function moveProperty(root, objectPath, propertyName, direction) {
    return mapValueAt(root, objectPath, (value) => {
        if (value.tag === 'jv-object') {
            const propIndex = value.properties.findIndex((p) => p.name === propertyName);
            const newProps = moveElems(value.properties, propIndex, direction);
            const newObject = Object.assign(Object.assign({}, value), { properties: newProps });
            return (0, tea_cup_core_1.just)(newObject);
        }
        return (0, tea_cup_core_1.just)(value);
    }).withDefault(root);
}
exports.moveProperty = moveProperty;
function moveElement(root, arrayPath, elemIndex, direction) {
    return mapValueAt(root, arrayPath, (value) => {
        if (value.tag === 'jv-array') {
            const newElems = moveElems(value.elems, elemIndex, direction);
            const newArray = Object.assign(Object.assign({}, value), { elems: newElems });
            return (0, tea_cup_core_1.just)(newArray);
        }
        return (0, tea_cup_core_1.just)(value);
    }).withDefault(root);
}
exports.moveElement = moveElement;
function mergeProperties(from, into) {
    // keep set of prop names to avoid overwriting existing stuff in dst
    const dstPropNames = new Set(into.properties.map((p) => p.name));
    const newProps = new Array(...into.properties);
    from.properties.forEach((p) => {
        if (!dstPropNames.has(p.name)) {
            newProps.push(p);
        }
    });
    return jvObject(newProps);
}
exports.mergeProperties = mergeProperties;
function indexOfPathInParent(root, path) {
    return path
        .parent()
        .andThen((parentPath) => path.lastElem().andThen((lastPathElem) => getValueAt(root, parentPath).map((parentValue) => {
        switch (parentValue.tag) {
            case 'jv-object': {
                return parentValue.properties.findIndex((p) => p.name === lastPathElem);
            }
            case 'jv-array': {
                const i = parseInt(lastPathElem);
                if (isNaN(i)) {
                    return -1;
                }
                return i;
            }
            default:
                return -1;
        }
    })))
        .withDefault(-1);
}
exports.indexOfPathInParent = indexOfPathInParent;
function clearPropertiesIfObject(v) {
    if (v.tag === 'jv-object') {
        return jvObject();
    }
    return v;
}
exports.clearPropertiesIfObject = clearPropertiesIfObject;
