// @diesel-parser/json-form/dist/JsonEditor.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonEditor = exports.subscriptions = exports.setDebounceMsPort = exports.setStrictModePort = exports.sendJsonPort = exports.setStrictMode = exports.update = exports.ViewJsonEditor = exports.init = void 0;
const JsFacade = __importStar(require("@diesel-parser/json-schema-facade-ts"));
const react_1 = __importDefault(require("react"));
const react_tea_cup_1 = require("react-tea-cup");
const tea_cup_core_1 = require("tea-cup-core");
const TPM = __importStar(require("tea-pop-menu"));
const Actions_1 = require("./Actions");
const ContextMenu_1 = require("./ContextMenu");
const ContextMenuRenderer_1 = require("./ContextMenuRenderer");
const JsonValue_1 = require("./JsonValue");
const JsPath_1 = require("./JsPath");
const Model_1 = require("./Model");
const Msg_1 = require("./Msg");
const OutMsg_1 = require("./OutMsg");
const Renderer_1 = require("./renderer/Renderer");
function init(language, schema, initialValue, strictMode, rendererFactory, debounceMs) {
    JsFacade.setLang(language);
    const model = (0, Model_1.initialModel)(language, schema, initialValue, strictMode, debounceMs);
    return reInitRenderers(model, rendererFactory);
}
exports.init = init;
function reInitRenderers(model, customRendererFactory) {
    return model.validationResult
        .map((validationResult) => {
        const renderers = JsFacade.getRenderers(validationResult);
        const newCustomRenderers = new Map();
        const cmds = [];
        for (const [path, rendererDef] of renderers) {
            if (rendererDef !== undefined) {
                const key = rendererDef.key;
                const renderer = customRendererFactory.getRenderer(key);
                if (renderer.type === 'Just') {
                    const existingModel = model.customRenderers.get(path);
                    const m = existingModel === undefined ? tea_cup_core_1.nothing : existingModel;
                    const jValue = (0, JsonValue_1.getValueAt)(model.root.b, JsPath_1.JsPath.parse(path));
                    if (jValue.type === 'Just') {
                        const mac = renderer.value.reinit({
                            path: JsPath_1.JsPath.parse(path),
                            formModel: model,
                            value: jValue.value,
                            model: m.map((x) => x.rendererModel),
                            schema: rendererDef.schemaValue,
                        });
                        const customRendererModel = {
                            rendererModel: mac[0],
                            key,
                        };
                        newCustomRenderers.set(path, (0, tea_cup_core_1.just)(customRendererModel));
                        const cmd = mac[1].map((msg) => {
                            return {
                                tag: 'renderer-child-msg',
                                path,
                                msg,
                            };
                        });
                        cmds.push(cmd);
                    }
                }
            }
        }
        const newModel = Object.assign(Object.assign({}, model), { customRenderers: newCustomRenderers });
        return tea_cup_core_1.Tuple.t2n(newModel, tea_cup_core_1.Cmd.batch(cmds));
    })
        .withDefaultSupply(() => (0, tea_cup_core_1.noCmd)(model));
}
function ViewJsonEditor(props) {
    const { model, dispatch, renderOptions } = props;
    return (react_1.default.createElement("div", { className: "diesel-json-editor" },
        react_1.default.createElement("div", { className: "diesel-json-editor-scrollpane" },
            !(renderOptions === null || renderOptions === void 0 ? void 0 : renderOptions.hideDocRoot) && (react_1.default.createElement("div", { className: 'doc-root' },
                react_1.default.createElement("em", null, model.t('documentRoot')),
                react_1.default.createElement(Renderer_1.ArrayCounter, { value: model.root.b }),
                react_1.default.createElement(Renderer_1.MenuTrigger, { dispatch: dispatch, path: JsPath_1.JsPath.empty, disabled: model.adding.isJust(), t: model.t, renderOptions: renderOptions }))),
            react_1.default.createElement(Renderer_1.ViewJsonValue, { model: model, path: JsPath_1.JsPath.empty, value: model.root.b, dispatch: dispatch, rendererFactory: props.rendererFactory, language: props.model.lang, renderOptions: renderOptions })),
        model.menuModel
            .map((menuModel) => (react_1.default.createElement("div", { className: 'diesel-json-editor-menu', key: 'editor-menu' },
            react_1.default.createElement(TPM.ViewMenu, { model: menuModel, dispatch: (0, tea_cup_core_1.map)(dispatch, Msg_1.contextMenuMsg), renderer: (0, ContextMenuRenderer_1.contextMenuRenderer)(model.t) }))))
            .withDefault(react_1.default.createElement(react_1.default.Fragment, null)),
        react_1.default.createElement("textarea", { id: 'dummy-textarea', "aria-label": 'hidden textarea' })));
}
exports.ViewJsonEditor = ViewJsonEditor;
function withOut(mac, outMsg) {
    return [mac[0], mac[1], outMsg];
}
function noOut(mac) {
    return withOut(mac, tea_cup_core_1.nothing);
}
function withOutValueChanged(prevModel, mac) {
    const prev = JSON.stringify((0, JsonValue_1.valueToAny)(prevModel.root.b));
    const cur = JSON.stringify((0, JsonValue_1.valueToAny)(mac[0].root.b));
    return [
        mac[0],
        mac[1],
        cur === prev ? tea_cup_core_1.nothing : (0, tea_cup_core_1.just)((0, OutMsg_1.outValueChanged)(mac[0].root.b)),
    ];
}
function update(msg, model, rendererFactory, menuFilter) {
    switch (msg.tag) {
        case 'delete-property':
            return withOutValueChanged(model, (0, Actions_1.actionDeleteValue)(model, msg.path));
        case 'update-property': {
            const mac = (0, Actions_1.actionUpdateValue)(model, msg.path, msg.value);
            return withOutValueChanged(model, mac);
        }
        case 'add-property-clicked':
            return noOut((0, Actions_1.actionAddPropertyClicked)(model, msg.path));
        case 'new-property-name-changed':
            return noOut((0, tea_cup_core_1.noCmd)((0, Model_1.updateAddingPropertyName)(model, msg.value)));
        case 'new-property-name-key-down':
            switch (msg.key) {
                case 'Enter':
                    return withOutValueChanged(model, (0, Actions_1.actionConfirmAddProperty)(model, true));
                case 'Escape':
                    return withOutValueChanged(model, (0, Actions_1.actionConfirmAddProperty)(model, false));
                default:
                    return noOut((0, tea_cup_core_1.noCmd)(model));
            }
        case 'add-prop-ok-cancel-clicked': {
            return withOutValueChanged(model, (0, Actions_1.actionConfirmAddProperty)(model, msg.ok));
        }
        case 'menu-trigger-clicked': {
            return noOut((0, Actions_1.actionTriggerClicked)(model, msg.path, msg.refBox, menuFilter));
        }
        case 'menu-msg': {
            return withOutValueChanged(model, model.menuModel
                .map((menuModel) => {
                const mco = TPM.update(msg.child, menuModel);
                const newModel = Object.assign(Object.assign({}, model), { menuModel: (0, tea_cup_core_1.just)(mco[0]) });
                const cmd = mco[1].map(Msg_1.contextMenuMsg);
                const outMsg = mco[2];
                return outMsg
                    .map((out) => {
                    switch (out.tag) {
                        case 'request-close': {
                            const mac2 = closeMenu(newModel);
                            return tea_cup_core_1.Tuple.fromNative(mac2)
                                .mapSecond((c) => tea_cup_core_1.Cmd.batch([cmd, c]))
                                .toNative();
                        }
                        case 'item-selected': {
                            return tea_cup_core_1.Tuple.fromNative((0, tea_cup_core_1.updatePiped)(model, (m) => closeMenu(m), (m) => (0, ContextMenu_1.executeContextMenuAction)(m, out.data)))
                                .mapSecond((c) => tea_cup_core_1.Cmd.batch([cmd, c]))
                                .toNative();
                        }
                    }
                    return tea_cup_core_1.Tuple.t2n(newModel, cmd);
                })
                    .withDefaultSupply(() => tea_cup_core_1.Tuple.t2n(newModel, cmd));
            })
                .withDefaultSupply(() => (0, tea_cup_core_1.noCmd)(model)));
        }
        case 'add-elem-clicked':
            return withOutValueChanged(model, (0, Actions_1.actionAddElementToArray)(model, msg.path));
        case 'set-json-str': {
            return noOut(init(model.lang, msg.schema, msg.json, model.strictMode, rendererFactory, model.debounceMs));
        }
        case 'toggle-expand-collapse': {
            return noOut((0, Actions_1.actionToggleExpandCollapsePath)(model, msg.path));
        }
        case 'add-property-btn-clicked': {
            return withOutValueChanged(model, (0, Actions_1.actionAddProperty)(model, msg.path, msg.propertyName));
        }
        case 'no-op':
            return noOut((0, tea_cup_core_1.noCmd)(model));
        case 'set-strict-mode':
            return noOut((0, tea_cup_core_1.noCmd)(setStrictMode(model, msg.strictMode)));
        case 'set-debounce-ms':
            return noOut((0, tea_cup_core_1.noCmd)(Object.assign(Object.assign({}, model), { debounceMs: msg.debounceMs })));
        case 'recompute-metadata': {
            const newModel = (0, Model_1.computeAll)((0, Model_1.doValidate)(model));
            return withOutValueChanged(model, reInitRenderers(newModel, rendererFactory));
        }
        case 'renderer-child-msg': {
            const rendererModel = model.customRenderers.get(msg.path);
            if (rendererModel && rendererModel.type === 'Just') {
                const renderer = rendererFactory.getRenderer(rendererModel.value.key);
                if (renderer.type === 'Just') {
                    const maco = renderer.value.update(msg.msg, rendererModel.value.rendererModel);
                    const newCustomRenderers = new Map(model.customRenderers);
                    const newCustomRendererModel = Object.assign(Object.assign({}, rendererModel.value), { rendererModel: maco[0] });
                    newCustomRenderers.set(msg.path, (0, tea_cup_core_1.just)(newCustomRendererModel));
                    const newModel = Object.assign(Object.assign({}, model), { customRenderers: newCustomRenderers });
                    const cmd = maco[1].map((childMsg) => {
                        return {
                            tag: 'renderer-child-msg',
                            msg: childMsg,
                            path: msg.path,
                        };
                    });
                    const newValue = maco[2];
                    switch (newValue.type) {
                        case 'Nothing': {
                            return noOut(tea_cup_core_1.Tuple.t2n(newModel, cmd));
                        }
                        case 'Just': {
                            const mac2 = (0, Actions_1.actionUpdateValue)(newModel, JsPath_1.JsPath.parse(msg.path), newValue.value);
                            return withOutValueChanged(model, [
                                mac2[0],
                                tea_cup_core_1.Cmd.batch([cmd, mac2[1]]),
                            ]);
                        }
                    }
                }
            }
            return noOut((0, tea_cup_core_1.noCmd)(model));
        }
    }
}
exports.update = update;
function closeMenu(model) {
    return (0, tea_cup_core_1.noCmd)(Object.assign(Object.assign({}, model), { menuModel: tea_cup_core_1.nothing }));
}
function setStrictMode(model, strictMode) {
    return Object.assign(Object.assign({}, model), { strictMode });
}
exports.setStrictMode = setStrictMode;
// the ports allow to send Msgs to the update loop from the outside
// TODO not a const !
exports.sendJsonPort = new tea_cup_core_1.Port();
exports.setStrictModePort = new tea_cup_core_1.Port();
exports.setDebounceMsPort = new tea_cup_core_1.Port();
function subscriptions(model) {
    // the menu's subs
    const subMenu = model.menuModel
        .map((mm) => TPM.subscriptions(mm).map(Msg_1.contextMenuMsg))
        .withDefaultSupply(() => tea_cup_core_1.Sub.none());
    // the ports subs
    const portSub = exports.sendJsonPort.subscribe(Msg_1.setJsonStr);
    const setStrictModePortSub = exports.setStrictModePort.subscribe(Msg_1.setStrictModeMsg);
    const setDebouceMsPortSub = exports.setDebounceMsPort.subscribe(Msg_1.setDebounceMsMsg);
    return tea_cup_core_1.Sub.batch([
        subMenu,
        portSub,
        setStrictModePortSub,
        setDebouceMsPortSub,
    ]);
}
exports.subscriptions = subscriptions;
function JsonEditor(props) {
    return (react_1.default.createElement(react_tea_cup_1.Program, { init: () => init(props.language, props.schema, props.value, props.strictMode, props.rendererFactory, props.debounceMs || 500), view: (dispatch, model) => (react_1.default.createElement(ViewJsonEditor, { dispatch: dispatch, model: model, rendererFactory: props.rendererFactory, renderOptions: props.renderOptions })), update: (msg, model) => {
            const maco = update(msg, model, props.rendererFactory, props.menuFilter);
            maco[2].forEach((outMsg) => {
                switch (outMsg.tag) {
                    case 'value-changed': {
                        props.onChange && props.onChange(outMsg.value);
                    }
                }
            });
            return [maco[0], maco[1]];
        }, subscriptions: subscriptions }));
}
exports.JsonEditor = JsonEditor;
