// @diesel-parser/json-form/dist/JsonValue.test.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const JsonValue_1 = require("./JsonValue");
const tea_cup_core_1 = require("tea-cup-core");
const JsPath_1 = require("./JsPath");
describe('JsonValue', () => {
    test('should parse from str', () => {
        expect((0, JsonValue_1.parseJsonValue)('null')).toEqual((0, tea_cup_core_1.ok)(JsonValue_1.jvNull));
        expect((0, JsonValue_1.parseJsonValue)('123')).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvNumber)('123')));
        expect((0, JsonValue_1.parseJsonValue)('true')).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvBool)(true)));
        expect((0, JsonValue_1.parseJsonValue)('"yalla"')).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvString)('yalla')));
        expect((0, JsonValue_1.parseJsonValue)(JSON.stringify({ foo: 'bar' }))).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvObject)([{ name: 'foo', value: (0, JsonValue_1.jvString)('bar') }])));
        expect((0, JsonValue_1.parseJsonValue)(JSON.stringify([1, 2, 3]))).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('1'), (0, JsonValue_1.jvNumber)('2'), (0, JsonValue_1.jvNumber)('3')])));
    });
    describe('should convert', () => {
        test('from any', () => {
            expect((0, JsonValue_1.valueFromAny)(null)).toEqual((0, tea_cup_core_1.ok)(JsonValue_1.jvNull));
            expect((0, JsonValue_1.valueFromAny)(123)).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvNumber)('123')));
            expect((0, JsonValue_1.valueFromAny)(true)).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvBool)(true)));
            expect((0, JsonValue_1.valueFromAny)('yalla')).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvString)('yalla')));
            expect((0, JsonValue_1.valueFromAny)({ foo: 'bar' })).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvObject)([{ name: 'foo', value: (0, JsonValue_1.jvString)('bar') }])));
            expect((0, JsonValue_1.valueFromAny)([1, { foo: 'bar' }, [2, 3]])).toEqual((0, tea_cup_core_1.ok)((0, JsonValue_1.jvArray)([
                (0, JsonValue_1.jvNumber)('1'),
                (0, JsonValue_1.jvObject)([
                    {
                        name: 'foo',
                        value: (0, JsonValue_1.jvString)('bar'),
                    },
                ]),
                (0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('2'), (0, JsonValue_1.jvNumber)('3')]),
            ])));
        });
        test('invalid number converts to string', () => {
            expect((0, JsonValue_1.valueToAny)((0, JsonValue_1.jvNumber)('yalla'))).toEqual('yalla');
        });
        test('number roundtrip', () => {
            (0, JsonValue_1.valueFromAny)(123).match((v) => {
                expect(v).toEqual((0, JsonValue_1.jvNumber)('123'));
                const a = (0, JsonValue_1.valueToAny)(v);
                expect(a).toEqual(123);
            }, (err) => {
                fail(err);
            });
        });
        test('to any', () => {
            expect((0, JsonValue_1.valueToAny)(JsonValue_1.jvNull)).toEqual(null);
            expect((0, JsonValue_1.valueToAny)((0, JsonValue_1.jvNumber)('123'))).toEqual(123);
            expect((0, JsonValue_1.valueToAny)((0, JsonValue_1.jvString)('yalla'))).toEqual('yalla');
            expect((0, JsonValue_1.valueToAny)((0, JsonValue_1.jvBool)(true))).toEqual(true);
            expect((0, JsonValue_1.valueToAny)((0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('1'), (0, JsonValue_1.jvNumber)('2')]))).toEqual([
                1, 2,
            ]);
            expect((0, JsonValue_1.valueToAny)((0, JsonValue_1.jvObject)([
                { name: 'foo', value: (0, JsonValue_1.jvString)('bar') },
                {
                    name: 'blah',
                    value: (0, JsonValue_1.jvObject)([
                        {
                            name: 'baz',
                            value: (0, JsonValue_1.jvArray)([(0, JsonValue_1.jvBool)(true), (0, JsonValue_1.jvNumber)('999')]),
                        },
                    ]),
                },
            ]))).toEqual({
                foo: 'bar',
                blah: {
                    baz: [true, 999],
                },
            });
        });
    });
    function valueFromAnyThrow(value) {
        const v = (0, JsonValue_1.valueFromAny)(value);
        switch (v.tag) {
            case 'Err': {
                throw new Error(v.err);
            }
            case 'Ok': {
                return v.value;
            }
        }
    }
    describe('can be queried', () => {
        test('at root', () => {
            expect((0, JsonValue_1.getValueAt)(JsonValue_1.jvNull, JsPath_1.JsPath.empty)).toEqual((0, tea_cup_core_1.just)(JsonValue_1.jvNull));
        });
        test('1st level prop', () => {
            expect((0, JsonValue_1.getValueAt)(valueFromAnyThrow({
                foo: 'bar',
                blah: 1,
                zzz: true,
            }), JsPath_1.JsPath.parse('foo'))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvString)('bar')));
        });
        test('1st level array', () => {
            expect((0, JsonValue_1.getValueAt)(valueFromAnyThrow([1, true, 'yalla']), JsPath_1.JsPath.parse('2'))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvString)('yalla')));
        });
        test('deeply nested', () => {
            const myObj = valueFromAnyThrow({
                a: {
                    b: {
                        c: {
                            d: [
                                1,
                                {
                                    foo: 'bar',
                                },
                                2,
                            ],
                            d2: 'yolo',
                        },
                    },
                    b2: true,
                },
                a2: 'yalla',
            });
            expect((0, JsonValue_1.getValueAt)(myObj, JsPath_1.JsPath.parse('a/b/c'))).toEqual((0, tea_cup_core_1.just)(valueFromAnyThrow({
                d: [
                    1,
                    {
                        foo: 'bar',
                    },
                    2,
                ],
                d2: 'yolo',
            })));
            expect((0, JsonValue_1.getValueAt)(myObj, JsPath_1.JsPath.parse('a/b/c/d/1/foo'))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvString)('bar')));
        });
    });
    describe('can be updated', () => {
        test('at root', () => {
            expect((0, JsonValue_1.setValueAt)(JsonValue_1.jvNull, JsPath_1.JsPath.empty, (0, JsonValue_1.jvNumber)('123'))).toEqual((0, JsonValue_1.jvNumber)('123'));
        });
        test('1st level prop', () => {
            expect((0, JsonValue_1.setValueAt)((0, JsonValue_1.jvObject)([
                { name: 'foo', value: (0, JsonValue_1.jvNumber)('12') },
                { name: 'bar', value: (0, JsonValue_1.jvNumber)('13') },
            ]), JsPath_1.JsPath.empty.append('foo'), (0, JsonValue_1.jvBool)(true))).toEqual((0, JsonValue_1.jvObject)([
                { name: 'foo', value: (0, JsonValue_1.jvBool)(true) },
                { name: 'bar', value: (0, JsonValue_1.jvNumber)('13') },
            ]));
        });
        test('1st level array', () => {
            expect((0, JsonValue_1.setValueAt)((0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('12'), (0, JsonValue_1.jvNumber)('13'), (0, JsonValue_1.jvNumber)('14')]), JsPath_1.JsPath.empty.append(1), (0, JsonValue_1.jvString)('yalla'))).toEqual((0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('12'), (0, JsonValue_1.jvString)('yalla'), (0, JsonValue_1.jvNumber)('14')]));
        });
        test('deeply nested', () => {
            const myObj = valueFromAnyThrow({
                a: {
                    b: {
                        c: {
                            d: [
                                1,
                                {
                                    foo: 'bar',
                                },
                                2,
                            ],
                        },
                    },
                    b2: true,
                },
                a2: 'yalla',
            });
            const expected = valueFromAnyThrow({
                a: {
                    b: {
                        c: {
                            d: [
                                1,
                                {
                                    foo: 999,
                                },
                                2,
                            ],
                        },
                    },
                    b2: true,
                },
                a2: 'yalla',
            });
            expect((0, JsonValue_1.setValueAt)(myObj, JsPath_1.JsPath.parse('a/b/c/d/1/foo'), (0, JsonValue_1.jvNumber)('999'))).toEqual(expected);
        });
    });
    describe('can be mapped', () => {
        test('root', () => {
            expect((0, JsonValue_1.mapValueAt)(JsonValue_1.jvNull, JsPath_1.JsPath.empty, () => (0, tea_cup_core_1.just)((0, JsonValue_1.jvNumber)('123')))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvNumber)('123')));
        });
        test('root delete', () => {
            expect((0, JsonValue_1.deleteValueAt)(JsonValue_1.jvNull, JsPath_1.JsPath.empty)).toEqual(tea_cup_core_1.nothing);
        });
        const root1 = (0, JsonValue_1.jvObject)([
            { name: 'foo', value: (0, JsonValue_1.jvString)('bar') },
            { name: 'blah', value: (0, JsonValue_1.jvString)('yalla') },
        ]);
        test('1st level prop', () => {
            expect((0, JsonValue_1.mapValueAt)(root1, JsPath_1.JsPath.parse('foo'), () => (0, tea_cup_core_1.just)((0, JsonValue_1.jvNumber)('123')))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvObject)([
                { name: 'foo', value: (0, JsonValue_1.jvNumber)('123') },
                { name: 'blah', value: (0, JsonValue_1.jvString)('yalla') },
            ])));
        });
        test('1st level prop delete', () => {
            expect((0, JsonValue_1.deleteValueAt)(root1, JsPath_1.JsPath.parse('foo'))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvObject)([{ name: 'blah', value: (0, JsonValue_1.jvString)('yalla') }])));
        });
        const root2 = (0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('123'), (0, JsonValue_1.jvString)('foo'), JsonValue_1.jvNull]);
        test('1st level array', () => {
            expect((0, JsonValue_1.mapValueAt)(root2, JsPath_1.JsPath.parse('1'), () => (0, tea_cup_core_1.just)((0, JsonValue_1.jvBool)(true)))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('123'), (0, JsonValue_1.jvBool)(true), JsonValue_1.jvNull])));
        });
        test('1st level array delete', () => {
            expect((0, JsonValue_1.deleteValueAt)(root2, JsPath_1.JsPath.parse('1'))).toEqual((0, tea_cup_core_1.just)((0, JsonValue_1.jvArray)([(0, JsonValue_1.jvNumber)('123'), JsonValue_1.jvNull])));
        });
        const root3 = valueFromAnyThrow({
            a: {
                b: {
                    c: {
                        d: [
                            1,
                            {
                                foo: 'bar',
                            },
                            2,
                        ],
                    },
                },
                b2: true,
            },
            a2: 'yalla',
        });
        test('deeply nested', () => {
            const mapped = (0, JsonValue_1.mapValueAt)(root3, JsPath_1.JsPath.parse('a/b/c/d/1/foo'), () => (0, tea_cup_core_1.just)(JsonValue_1.jvNull));
            expect(mapped).toEqual((0, tea_cup_core_1.just)(valueFromAnyThrow({
                a: {
                    b: {
                        c: {
                            d: [
                                1,
                                {
                                    foo: null,
                                },
                                2,
                            ],
                        },
                    },
                    b2: true,
                },
                a2: 'yalla',
            })));
        });
        test('deeply nested delete', () => {
            expect((0, JsonValue_1.deleteValueAt)(root3, JsPath_1.JsPath.parse('a/b/c/d/1/foo'))).toEqual((0, tea_cup_core_1.just)(valueFromAnyThrow({
                a: {
                    b: {
                        c: {
                            d: [1, {}, 2],
                        },
                    },
                    b2: true,
                },
                a2: 'yalla',
            })));
        });
    });
    describe('should move', () => {
        test('props', () => {
            const v = valueFromAnyThrow({
                a: 1,
                b: 2,
                c: 3,
            });
            expect((0, JsonValue_1.moveProperty)(v, JsPath_1.JsPath.empty, 'b', 'up')).toEqual(valueFromAnyThrow({
                b: 2,
                a: 1,
                c: 3,
            }));
            expect((0, JsonValue_1.moveProperty)(v, JsPath_1.JsPath.empty, 'b', 'down')).toEqual(valueFromAnyThrow({
                a: 1,
                c: 3,
                b: 2,
            }));
            expect((0, JsonValue_1.moveProperty)(v, JsPath_1.JsPath.empty, 'a', 'up')).toEqual(valueFromAnyThrow({
                a: 1,
                b: 2,
                c: 3,
            }));
            expect((0, JsonValue_1.moveProperty)(v, JsPath_1.JsPath.empty, 'c', 'down')).toEqual(valueFromAnyThrow({
                a: 1,
                b: 2,
                c: 3,
            }));
            expect((0, JsonValue_1.moveProperty)(v, JsPath_1.JsPath.empty, 'c', 'up')).toEqual(valueFromAnyThrow({
                a: 1,
                c: 3,
                b: 2,
            }));
            expect((0, JsonValue_1.moveProperty)(v, JsPath_1.JsPath.empty, 'not-existing', 'up')).toEqual(v);
        });
        test('arrays', () => {
            const v = valueFromAnyThrow([1, 2, 3]);
            expect((0, JsonValue_1.moveElement)(v, JsPath_1.JsPath.empty, 1, 'up')).toEqual(valueFromAnyThrow([2, 1, 3]));
            expect((0, JsonValue_1.moveElement)(v, JsPath_1.JsPath.empty, 1, 'down')).toEqual(valueFromAnyThrow([1, 3, 2]));
            expect((0, JsonValue_1.moveElement)(v, JsPath_1.JsPath.empty, 0, 'up')).toEqual(valueFromAnyThrow([1, 2, 3]));
            expect((0, JsonValue_1.moveElement)(v, JsPath_1.JsPath.empty, 2, 'down')).toEqual(valueFromAnyThrow([1, 2, 3]));
            expect((0, JsonValue_1.moveElement)(v, JsPath_1.JsPath.empty, 999, 'up')).toEqual(v);
        });
    });
    test('objects can be merged', () => {
        const o1 = valueFromAnyThrow({
            a: 1,
            c: 3,
        });
        const o2 = valueFromAnyThrow({
            a: 2,
            b: 2,
        });
        expect((0, JsonValue_1.mergeProperties)(o1, o2)).toEqual(valueFromAnyThrow({
            a: 2,
            b: 2,
            c: 3,
        }));
        expect((0, JsonValue_1.mergeProperties)(o2, o1)).toEqual(valueFromAnyThrow({
            a: 1,
            c: 3,
            b: 2,
        }));
    });
});
