// @diesel-parser/json-form/dist/JsPath.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsPath = void 0;
const tea_cup_core_1 = require("tea-cup-core");
class JsPath {
    constructor(_elems) {
        this._elems = _elems;
    }
    isEmpty() {
        return this._elems.length === 0;
    }
    get elems() {
        return this._elems;
    }
    head() {
        return (0, tea_cup_core_1.maybeOf)(this.elems[0]);
    }
    tail() {
        const newElems = this._elems.slice(1);
        return new JsPath(newElems);
    }
    format(separator = '/') {
        return this._elems.join(separator);
    }
    equals(other) {
        return this.format() === other.format();
    }
    append(elem) {
        return new JsPath(this._elems.concat([elem.toString()]));
    }
    parent() {
        if (this._elems.length === 0) {
            return tea_cup_core_1.nothing;
        }
        const parts = this._elems.slice(0, -1);
        return (0, tea_cup_core_1.just)(new JsPath(parts));
    }
    lastElem() {
        return (0, tea_cup_core_1.maybeOf)(this._elems[this._elems.length - 1]);
    }
    isParentOf(child) {
        return child.format().startsWith(this.format());
    }
    static parse(path) {
        if (path === '') {
            return JsPath.empty;
        }
        const parts = path.split('/');
        return new JsPath(parts);
    }
}
exports.JsPath = JsPath;
JsPath.empty = new JsPath([]);
