// @diesel-parser/json-form/dist/Model.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFormats = exports.getProposals = exports.updateAddingPropertyName = exports.computeAll = exports.initialModel = exports.computeStringsMetadata = exports.computePropertiesToAdd = exports.computeErrors = exports.doValidate = void 0;
const tea_cup_core_1 = require("tea-cup-core");
const JsonValue_1 = require("./JsonValue");
const JsPath_1 = require("./JsPath");
const JsFacade = __importStar(require("@diesel-parser/json-schema-facade-ts"));
const MyI18n_1 = require("./i18n/MyI18n");
function doValidate(model) {
    return model.schema
        .map((t) => {
        const validationResult = (0, tea_cup_core_1.just)(JsFacade.validate(t.a, model.root.a));
        return Object.assign(Object.assign({}, model), { validationResult });
    })
        .withDefault(model);
}
exports.doValidate = doValidate;
function computeErrors(model) {
    const errors = model.validationResult
        .map(JsFacade.getErrors)
        .map((jsErrors) => {
        const res = new Map();
        jsErrors.forEach((err) => {
            const errsAtPath = res.get(err.path);
            if (errsAtPath) {
                res.set(err.path, [...errsAtPath, err]);
            }
            else {
                res.set(err.path, [err]);
            }
        });
        return res;
    })
        .withDefaultSupply(() => new Map());
    return Object.assign(Object.assign({}, model), { errors });
}
exports.computeErrors = computeErrors;
function computePropertiesToAdd(model) {
    return model.validationResult
        .map((validationResult) => {
        const propertiesToAdd = new Map();
        doComputePropsToAdd(model, validationResult, propertiesToAdd);
        const newModel = Object.assign(Object.assign({}, model), { propertiesToAdd });
        return newModel;
    })
        .withDefault(model);
}
exports.computePropertiesToAdd = computePropertiesToAdd;
function computeStringsMetadata(model) {
    return model.validationResult
        .map((validationResult) => {
        const comboBoxes = new Map();
        const formats = new Map();
        const metadata = {
            comboBoxes,
            formats,
        };
        doComputeStringsMetadata(model, validationResult, metadata);
        const newModel = Object.assign(Object.assign({}, model), { comboBoxes,
            formats });
        return newModel;
    })
        .withDefault(model);
}
exports.computeStringsMetadata = computeStringsMetadata;
function doComputeStringsMetadata(model, validationResult, metadata, path = JsPath_1.JsPath.empty) {
    (0, JsonValue_1.getValueAt)(model.root.b, path).forEach((value) => {
        switch (value.tag) {
            case 'jv-object': {
                value.properties.forEach((prop) => doComputeStringsMetadata(model, validationResult, metadata, path.append(prop.name)));
                break;
            }
            case 'jv-array': {
                // recurse
                value.elems.forEach((elem, elemIndex) => doComputeStringsMetadata(model, validationResult, metadata, path.append(elemIndex)));
                break;
            }
            case 'jv-string': {
                const proposals = getProposals(validationResult, path, -1)
                    .flatMap((proposal) => {
                    if (proposal.tag === 'jv-string') {
                        return [proposal.value];
                    }
                    else {
                        return [];
                    }
                })
                    .filter((s) => s !== '');
                if (proposals.length > 0) {
                    metadata.comboBoxes.set(path.format(), proposals);
                }
                const formats = getFormats(validationResult, path);
                if (formats.length > 0) {
                    metadata.formats.set(path.format(), formats);
                }
            }
        }
    });
}
function doComputePropsToAdd(model, validationResult, props, path = JsPath_1.JsPath.empty) {
    (0, JsonValue_1.getValueAt)(model.root.b, path).forEach((value) => {
        switch (value.tag) {
            case 'jv-object': {
                // compute props for this object and recurse
                const propNameProposals = getProposals(validationResult, path, -1).flatMap((proposal) => {
                    if (proposal.tag === 'jv-object') {
                        return proposal.properties.map((p) => p.name);
                    }
                    return [];
                });
                props.set(path.format(), propNameProposals);
                value.properties.forEach((prop) => doComputePropsToAdd(model, validationResult, props, path.append(prop.name)));
                break;
            }
            case 'jv-array': {
                // recurse
                value.elems.forEach((elem, elemIndex) => doComputePropsToAdd(model, validationResult, props, path.append(elemIndex)));
                break;
            }
        }
    });
}
function toValueTuple(v) {
    return new tea_cup_core_1.Tuple((0, JsonValue_1.valueToAny)(v), v);
}
function initialModel(lang, schema, root, strictMode, debounceMs) {
    const t = (0, MyI18n_1.initMyI18n)(lang);
    const model = {
        lang,
        t,
        schema: schema.map(toValueTuple),
        root: toValueTuple(root),
        validationResult: tea_cup_core_1.nothing,
        errors: new Map(),
        adding: tea_cup_core_1.nothing,
        menuModel: tea_cup_core_1.nothing,
        collapsedPaths: new Set(),
        propertiesToAdd: new Map(),
        comboBoxes: new Map(),
        formats: new Map(),
        strictMode,
        customRenderers: new Map(),
        debounceMs: Math.max(0, debounceMs),
    };
    return computeAll(doValidate(model));
}
exports.initialModel = initialModel;
function computeAll(model) {
    return computeStringsMetadata(computePropertiesToAdd(computeErrors(model)));
}
exports.computeAll = computeAll;
function updateAddingPropertyName(model, propName) {
    return Object.assign(Object.assign({}, model), { adding: model.adding.map((addingState) => {
            const res = Object.assign(Object.assign({}, addingState), { addingPropName: propName, isDuplicate: (0, JsonValue_1.getValueAt)(model.root.b, addingState.ownerPath)
                    .map((ownerValue) => {
                    if (ownerValue.tag === 'jv-object') {
                        return (ownerValue.properties.find((p) => p.name === propName) !==
                            undefined);
                    }
                    else {
                        return false;
                    }
                })
                    .withDefault(false) });
            return res;
        }) });
}
exports.updateAddingPropertyName = updateAddingPropertyName;
function getProposals(validationResult, path, depth) {
    const proposals = JsFacade.propose(validationResult, path.format(), depth);
    return proposals.flatMap((proposalAny) => {
        return (0, JsonValue_1.valueFromAny)(proposalAny).match((jsonValue) => [jsonValue], () => []);
    });
}
exports.getProposals = getProposals;
function getFormats(validationResult, path) {
    return JsFacade.getFormats(validationResult, path.format());
}
exports.getFormats = getFormats;
