// @diesel-parser/json-form/dist/Msg.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.noOp = exports.contextMenuMsg = exports.setDebounceMsMsg = exports.setStrictModeMsg = exports.setJsonStr = void 0;
function setJsonStr(schemaAndJson) {
    return {
        tag: 'set-json-str',
        schema: schemaAndJson[0],
        json: schemaAndJson[1],
    };
}
exports.setJsonStr = setJsonStr;
function setStrictModeMsg(strictMode) {
    return {
        tag: 'set-strict-mode',
        strictMode,
    };
}
exports.setStrictModeMsg = setStrictModeMsg;
function setDebounceMsMsg(debounceMs) {
    return {
        tag: 'set-debounce-ms',
        debounceMs,
    };
}
exports.setDebounceMsMsg = setDebounceMsMsg;
function contextMenuMsg(child) {
    return {
        tag: 'menu-msg',
        child,
    };
}
exports.contextMenuMsg = contextMenuMsg;
exports.noOp = {
    tag: 'no-op',
};
