// @diesel-parser/json-form/dist/RenderOptions.js
"use strict";
/*
 * Copyright 2018, 2024 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMenuOptionHidden = exports.MenuOptions = void 0;
var MenuOptions;
(function (MenuOptions) {
    MenuOptions[MenuOptions["MOVE"] = 1] = "MOVE";
    MenuOptions[MenuOptions["DELETE"] = 2] = "DELETE";
    MenuOptions[MenuOptions["PROPOSE"] = 4] = "PROPOSE";
    MenuOptions[MenuOptions["CHANGE_TYPE"] = 8] = "CHANGE_TYPE";
    MenuOptions[MenuOptions["ADD"] = 16] = "ADD";
})(MenuOptions = exports.MenuOptions || (exports.MenuOptions = {}));
function isMenuOptionHidden(option, filter) {
    var _a;
    return (((_a = filter === null || filter === void 0 ? void 0 : filter.menuFilters) !== null && _a !== void 0 ? _a : 0) & option) > 0;
}
exports.isMenuOptionHidden = isMenuOptionHidden;
