// @diesel-parser/json-form/dist/RenderOptions.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright 2018, 2024 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const react_1 = require("@testing-library/react");
require("@testing-library/jest-dom");
const JsPath_1 = require("./JsPath");
const Renderer_1 = require("./renderer/Renderer");
const JsonEditor_1 = require("./JsonEditor");
const Model_1 = require("./Model");
const tea_cup_core_1 = require("tea-cup-core");
const JsonValue_1 = require("./JsonValue");
describe('Render options', () => {
    describe('defaults', () => {
        test('menu tooltip', () => {
            const { container } = (0, react_1.render)((0, Renderer_1.MenuTrigger)({
                path: JsPath_1.JsPath.empty,
                disabled: false,
                t: () => 'no-t',
                dispatch: () => { },
                renderOptions: {},
            }));
            expect(container.querySelector('button')).not.toHaveAttribute('style');
        });
        test('menu document root', () => {
            const { container } = (0, react_1.render)((0, JsonEditor_1.ViewJsonEditor)({
                model: (0, Model_1.initialModel)('en', tea_cup_core_1.nothing, JsonValue_1.jvNull, false, 0),
                dispatch: () => { },
                rendererFactory: new Renderer_1.RendererFactory(),
                renderOptions: {},
            }));
            expect(container.querySelectorAll('.doc-root')).toHaveLength(1);
        });
        test('menu collapsible panel in object', () => {
            const { container } = (0, react_1.render)((0, Renderer_1.ViewJsonValue)({
                model: (0, Model_1.initialModel)('en', tea_cup_core_1.nothing, JsonValue_1.jvNull, false, 0),
                path: JsPath_1.JsPath.empty,
                value: (0, JsonValue_1.jvObject)([{ name: 'foo', value: JsonValue_1.jvNull }]),
                language: 'en',
                dispatch: () => { },
                rendererFactory: new Renderer_1.RendererFactory(),
                renderOptions: {},
            }));
            expect(container.querySelectorAll('.prop-expand')).toHaveLength(1);
        });
        test('menu collapsible panel in array', () => {
            const { container } = (0, react_1.render)((0, Renderer_1.ViewJsonValue)({
                model: (0, Model_1.initialModel)('en', tea_cup_core_1.nothing, JsonValue_1.jvNull, false, 0),
                path: JsPath_1.JsPath.empty,
                value: (0, JsonValue_1.jvArray)([JsonValue_1.jvNull]),
                language: 'en',
                dispatch: () => { },
                rendererFactory: new Renderer_1.RendererFactory(),
                renderOptions: {},
            }));
            expect(container.querySelectorAll('.prop-expand')).toHaveLength(1);
        });
    });
    describe('with options', () => {
        test('hide menu tooltip', () => {
            const { container } = (0, react_1.render)((0, Renderer_1.MenuTrigger)({
                path: JsPath_1.JsPath.empty,
                disabled: false,
                t: () => 'no-t',
                dispatch: () => { },
                renderOptions: { hideMenuTooltip: true },
            }));
            expect(container.querySelector('button')).toHaveAttribute('style', 'overflow: hidden;');
        });
        test('hide menu document root', () => {
            const { container } = (0, react_1.render)((0, JsonEditor_1.ViewJsonEditor)({
                model: (0, Model_1.initialModel)('en', tea_cup_core_1.nothing, JsonValue_1.jvNull, false, 0),
                dispatch: () => { },
                rendererFactory: new Renderer_1.RendererFactory(),
                renderOptions: { hideDocRoot: true },
            }));
            expect(container.querySelectorAll('.doc-root')).toHaveLength(0);
        });
        test('hide menu collapsible panel in object', () => {
            const { container } = (0, react_1.render)((0, Renderer_1.ViewJsonValue)({
                model: (0, Model_1.initialModel)('en', tea_cup_core_1.nothing, JsonValue_1.jvNull, false, 0),
                path: JsPath_1.JsPath.empty,
                value: (0, JsonValue_1.jvObject)([{ name: 'foo', value: JsonValue_1.jvNull }]),
                language: 'en',
                dispatch: () => { },
                rendererFactory: new Renderer_1.RendererFactory(),
                renderOptions: { hideCollapsiblePanel: true },
            }));
            expect(container.querySelectorAll('.prop-expand')).toHaveLength(0);
        });
        test('hide menu collapsible panel in array', () => {
            const { container } = (0, react_1.render)((0, Renderer_1.ViewJsonValue)({
                model: (0, Model_1.initialModel)('en', tea_cup_core_1.nothing, JsonValue_1.jvNull, false, 0),
                path: JsPath_1.JsPath.empty,
                value: (0, JsonValue_1.jvArray)([JsonValue_1.jvNull]),
                language: 'en',
                dispatch: () => { },
                rendererFactory: new Renderer_1.RendererFactory(),
                renderOptions: { hideCollapsiblePanel: true },
            }));
            expect(container.querySelectorAll('.prop-expand')).toHaveLength(0);
        });
    });
});
