// @diesel-parser/json-form/dist/renderer/Renderer.js
"use strict";
/*
 * Copyright 2018 The Diesel Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuTrigger = exports.WrapErrors = exports.ViewErrors = exports.onMenuTriggerClick = exports.ArrayCounter = exports.ViewJsonValue = exports.RendererFactory = void 0;
const carbon_components_react_1 = require("carbon-components-react");
const React = __importStar(require("react"));
const tea_cup_core_1 = require("tea-cup-core");
const tea_pop_core_1 = require("tea-pop-core");
const JsonValue_1 = require("../JsonValue");
const _16_1 = __importDefault(require("@carbon/icons-react/lib/add/16"));
const _16_2 = __importDefault(require("@carbon/icons-react/lib/chevron--down/16"));
const _16_3 = __importDefault(require("@carbon/icons-react/lib/chevron--up/16"));
const _16_4 = __importDefault(require("@carbon/icons-react/lib/overflow-menu--vertical/16"));
class RendererFactory {
    constructor() {
        this.renderers = new Map();
    }
    addRenderer(key, renderer) {
        this.renderers.set(key, renderer);
    }
    getRenderer(key) {
        const renderer = this.renderers.get(key);
        return (0, tea_cup_core_1.maybeOf)(renderer);
    }
}
exports.RendererFactory = RendererFactory;
function ViewJsonValue(p) {
    const { value, rendererFactory } = p;
    const path = p.path.format();
    if (rendererFactory) {
        const customRendererModel = p.model.customRenderers.get(path);
        if (customRendererModel && customRendererModel.type === 'Just') {
            const m = customRendererModel.value;
            const renderer = rendererFactory.getRenderer(m.key);
            if (renderer.type === 'Just') {
                return renderer.value.view({
                    dispatch: (msg) => p.dispatch({
                        tag: 'renderer-child-msg',
                        path,
                        msg,
                    }),
                    model: m.rendererModel,
                    path: p.path,
                    formView: (path, value) => (React.createElement(ViewJsonValue, { model: p.model, path: path, value: value, dispatch: p.dispatch, rendererFactory: rendererFactory, language: p.language })),
                });
            }
        }
    }
    const content = () => {
        switch (value.tag) {
            case 'jv-array':
                return React.createElement(ViewArray, Object.assign({}, p, { value: value }));
            case 'jv-object':
                return React.createElement(ViewObject, Object.assign({}, p, { value: value }));
            case 'jv-boolean':
                return React.createElement(ViewBoolean, Object.assign({}, p, { value: value }));
            case 'jv-null':
                return React.createElement(ViewNull, Object.assign({}, p, { value: value }));
            case 'jv-number':
                return React.createElement(ViewNumber, Object.assign({}, p, { value: value }));
            case 'jv-string':
                return React.createElement(ViewString, Object.assign({}, p, { value: value }));
        }
    };
    return (React.createElement("div", { className: 'value', "data-path": p.path.format() }, content()));
}
exports.ViewJsonValue = ViewJsonValue;
function ViewObject(p) {
    const { model, value, dispatch, path, renderOptions } = p;
    const { properties } = value;
    const { t } = model;
    const isAddingProp = p.model.adding.isJust();
    let addSection = React.createElement(React.Fragment, null);
    model.adding.forEach((addingState) => {
        // are we owning ?
        const { ownerPath } = addingState;
        if (path.equals(ownerPath)) {
            addSection = (React.createElement("div", { className: "add-prop-form" },
                React.createElement(carbon_components_react_1.TextInput, { labelText: t('propertyNameLabel', {
                        path: path.format('.'),
                    }).toString(), hideLabel: true, id: 'property-name-editor', placeholder: t('propertyNamePlaceholder'), value: addingState.addingPropName, onChange: (e) => dispatch({
                        tag: 'new-property-name-changed',
                        value: e.target.value,
                    }), onKeyDown: (e) => {
                        dispatch({
                            tag: 'new-property-name-key-down',
                            key: e.key,
                        });
                    }, invalidText: t('propertyAlreadyExists'), invalid: addingState.isDuplicate }),
                React.createElement("div", { className: 'buttons-row' },
                    React.createElement("div", { className: "spacer" }),
                    React.createElement(carbon_components_react_1.Button, { kind: 'primary', disabled: addingState.addingPropName === '' || addingState.isDuplicate, onClick: () => dispatch({ tag: 'add-prop-ok-cancel-clicked', ok: true }) }, "Add"),
                    React.createElement(carbon_components_react_1.Button, { kind: 'secondary', onClick: () => dispatch({ tag: 'add-prop-ok-cancel-clicked', ok: false }) }, "Cancel"))));
        }
    });
    const addingOwnerPath = model.adding.map((a) => a.ownerPath);
    const existingPropertyNames = new Set(properties.map((p) => p.name));
    return (React.createElement("div", { className: "jv-object" },
        properties.length === 0 ? (React.createElement("div", { className: "empty-obj" }, t('emptyObject'))) : (React.createElement(React.Fragment, null)),
        properties.map((prop, propIndex) => {
            const propertyPath = p.path.append(prop.name);
            const disabled = isAddingProp &&
                addingOwnerPath
                    .map((ownerPath) => {
                    return !propertyPath.isParentOf(ownerPath);
                })
                    .withDefault(false);
            const propNameClass = ['object-prop'].concat(disabled ? ['disabled'] : ['']);
            const isCollapsed = model.collapsedPaths.has(propertyPath.format());
            return (React.createElement("div", { className: propNameClass.join(' '), key: prop.name + propIndex },
                React.createElement("div", { className: 'prop-name-row' },
                    !(renderOptions === null || renderOptions === void 0 ? void 0 : renderOptions.hideCollapsiblePanel) && (React.createElement("div", { className: "prop-expand" },
                        React.createElement(ExpandCollapseButton, { collapsed: isCollapsed, dispatch: dispatch, path: propertyPath, t: t }))),
                    React.createElement("div", { className: 'prop-name' }, prop.name),
                    React.createElement(ArrayCounter, { value: prop.value }),
                    React.createElement("div", { className: 'prop-menu' },
                        React.createElement(MenuTrigger, { dispatch: dispatch, path: propertyPath, disabled: isAddingProp, t: t, renderOptions: renderOptions }))),
                isCollapsed ? (React.createElement(React.Fragment, null)) : (React.createElement("div", { className: "prop-value" },
                    React.createElement(ViewJsonValue, Object.assign({}, p, { path: propertyPath, value: prop.value, renderOptions: renderOptions }))))));
        }),
        React.createElement(ViewErrors, { errors: getErrorsAtPath(p) }),
        React.createElement("div", null, addSection),
        React.createElement("div", null, (0, tea_cup_core_1.maybeOf)(model.propertiesToAdd.get(p.path.format()))
            .map((propNames) => (React.createElement(React.Fragment, null, propNames
            .filter((propName) => !existingPropertyNames.has(propName))
            .sort()
            .map((propName) => (React.createElement("div", { className: "add-prop-row", key: propName },
            React.createElement(carbon_components_react_1.Button, { renderIcon: _16_1.default, kind: 'ghost', onClick: () => dispatch({
                    tag: 'add-property-btn-clicked',
                    path: p.path,
                    propertyName: propName,
                }) }, propName)))))))
            .withDefault(React.createElement(React.Fragment, null)))));
}
function ArrayCounter(props) {
    if (props.value.tag === 'jv-array') {
        return (React.createElement("div", { className: "array-counter" },
            React.createElement(carbon_components_react_1.Tag, null, props.value.elems.length)));
    }
    return React.createElement(React.Fragment, null);
}
exports.ArrayCounter = ArrayCounter;
function ExpandCollapseButton(props) {
    return (React.createElement(carbon_components_react_1.Button, { kind: 'ghost', renderIcon: props.collapsed ? _16_3.default : _16_2.default, iconDescription: props.t(props.collapsed ? 'icon.expand' : 'icon.collapse'), 
        // disabled={disabled}
        size: 'sm', tooltipPosition: 'right', hasIconOnly: true, onClick: () => props.dispatch({
            tag: 'toggle-expand-collapse',
            path: props.path,
        }) }));
}
function ViewArray(p) {
    const { dispatch, path, value, model, renderOptions } = p;
    const { t } = model;
    const isAdding = p.model.adding.isJust();
    return (React.createElement(exports.WrapErrors, Object.assign({}, p),
        React.createElement("div", { className: "jv-array" }, value.elems.length === 0 ? (React.createElement("div", { className: "empty-obj" }, t('emptyArray'))) : (React.createElement(React.Fragment, null, value.elems.map((elemValue, elemIndex) => {
            const elemPath = path.append(elemIndex.toString());
            const isCollapsed = p.model.collapsedPaths.has(elemPath.format());
            return (React.createElement("div", { className: 'array-elem', key: `value-${elemIndex}` },
                React.createElement("div", { className: 'array-elem-head' },
                    !(renderOptions === null || renderOptions === void 0 ? void 0 : renderOptions.hideCollapsiblePanel) && (React.createElement("div", { className: "prop-expand" },
                        React.createElement(ExpandCollapseButton, { collapsed: isCollapsed, dispatch: dispatch, path: elemPath, t: t }))),
                    React.createElement("div", { className: 'elem-name' },
                        "#",
                        elemIndex),
                    React.createElement("div", { className: 'prop-menu' },
                        React.createElement(MenuTrigger, { dispatch: dispatch, path: elemPath, disabled: isAdding, t: t, renderOptions: renderOptions }))),
                isCollapsed ? (React.createElement(React.Fragment, null)) : (React.createElement("div", { className: 'elem-value' },
                    React.createElement(ViewJsonValue, Object.assign({}, p, { path: elemPath, value: elemValue, renderOptions: renderOptions }))))));
        }))))));
}
function dispatchUpdateProperty(p, newValue) {
    p.dispatch({
        tag: 'update-property',
        path: p.path,
        value: newValue,
    });
}
function ViewNumber(p) {
    const { t } = p.model;
    return (React.createElement(carbon_components_react_1.TextInput, { labelText: t('numberValueLabel', { path: p.path.format('.') }).toString(), hideLabel: true, id: 'input-' + p.path.format('_'), value: p.value.value, disabled: p.model.adding.isJust(), invalidText: errorsToInvalidText(p), invalid: isInvalid(p), onChange: (evt) => {
            dispatchUpdateProperty(p, {
                tag: 'jv-number',
                value: evt.target.value,
            });
        } }));
}
function errorsToInvalidText(p) {
    return getErrorsAtPath(p)
        .map((e) => e.message)
        .join(', ');
}
function getErrorsAtPath(p) {
    var _a;
    return (_a = p.model.errors.get(p.path.format())) !== null && _a !== void 0 ? _a : [];
}
function isInvalid(p) {
    return getErrorsAtPath(p).length > 0;
}
function ViewStringDefault(p) {
    const { t } = p.model;
    return (React.createElement(carbon_components_react_1.TextInput, { labelText: t('stringValueLabel', { path: p.path.format('.') }).toString(), hideLabel: true, id: 'input-' + p.path.format('_'), type: "text", value: p.value.value, disabled: p.model.adding.isJust(), invalidText: errorsToInvalidText(p), invalid: isInvalid(p), placeholder: t('stringValuePlaceholder'), onChange: (evt) => {
            const newStr = evt.target.value;
            dispatchUpdateProperty(p, { tag: 'jv-string', value: newStr });
        } }));
}
function ViewString(p) {
    var _a, _b;
    const fmtPath = p.path.format();
    const formats = (_a = p.model.formats.get(fmtPath)) !== null && _a !== void 0 ? _a : [];
    if (formats.length === 0) {
        // no formats, check for proposals (enum, examples etc)
        const proposals = (_b = p.model.comboBoxes.get(fmtPath)) !== null && _b !== void 0 ? _b : [];
        if (proposals.length === 0) {
            return React.createElement(ViewStringDefault, Object.assign({}, p));
        }
        else {
            return ViewStringWithCombo(Object.assign(Object.assign({}, p), { proposals }));
        }
    }
    else {
        // we have formats !
        return React.createElement(ViewStringWithFormats, Object.assign({}, p, { formats: formats }));
    }
}
function ViewStringWithCombo(p) {
    const { t } = p.model;
    return (React.createElement(carbon_components_react_1.ComboBox, { id: 'input-' + p.path.format('_'), ariaLabel: t('stringValueComboLabel', {
            path: p.path.format('.'),
        }).toString(), disabled: p.model.adding.isJust(), invalidText: errorsToInvalidText(p), invalid: isInvalid(p), items: p.proposals, value: p.value.value, selectedItem: p.value.value, placeholder: '', onChange: (item) => {
            var _a;
            dispatchUpdateProperty(p, {
                tag: 'jv-string',
                value: (_a = item.selectedItem) !== null && _a !== void 0 ? _a : '',
            });
        }, onInputChange: (text) => {
            dispatchUpdateProperty(p, {
                tag: 'jv-string',
                value: text !== null && text !== void 0 ? text : '',
            });
        } }));
}
function MyTimePicker(props) {
    const fmtPath = props.path.format('_');
    const { onChange, value, isInvalid, invalidText, t } = props;
    return (React.createElement(carbon_components_react_1.TimePicker, { id: 'time-picker-' + fmtPath, className: 'time-picker', onChange: (e) => {
            const t = new MyTime(value);
            const replaced = t.setTime(e.target.value);
            onChange(replaced.fullTime);
        }, pattern: "([0-1]\\d|2[0-3]):([0-5]\\d)(:[0-5]\\d(\\.(\\d{1,3}))?)?", maxLength: 12, invalidText: invalidText, invalid: isInvalid, value: new MyTime(value).time },
        React.createElement("div", { className: 'time-picker-offset-wrapper' },
            React.createElement(carbon_components_react_1.TimePickerSelect, { id: 'time-picker-select' + fmtPath, "aria-label": t('timeValueLabel', {
                    path: props.path.format('.'),
                }).toString(), labelText: '', onChange: (event) => {
                    const offset = event.target.value;
                    const t = new MyTime(value);
                    onChange(t.setOffset(offset).fullTime);
                }, value: new MyTime(value).offset }, [React.createElement(carbon_components_react_1.SelectItem, { text: '', value: '', key: 'item' })].concat(allOffsets.map((offset, i) => (React.createElement(carbon_components_react_1.SelectItem, { value: offset, text: offset, key: 'item-' + i }))))))));
}
function languageToPickerLocale(x) {
    const parts = x.split('-');
    // @ts-ignore
    return parts[0];
}
function MyDatePicker(props) {
    const fmtPath = props.path.format('_');
    const { t } = props;
    return (React.createElement(carbon_components_react_1.DatePicker, { id: 'date-picker-' + fmtPath, datePickerType: "single", dateFormat: 'Y-m-d', onChange: (dates, str) => {
            props.onChange(str);
        }, value: props.value, locale: languageToPickerLocale(props.language) },
        React.createElement(carbon_components_react_1.DatePickerInput, { id: 'input-' + fmtPath, "aria-label": t('dateValueLabel', {
                path: props.path.format('.'),
            }).toString(), labelText: '', hideLabel: true, autoComplete: 'off', value: props.value, onChange: (e) => {
                props.onChange(e.target.value);
            }, invalidText: props.invalidText, invalid: props.isInvalid })));
}
function ViewStringWithFormats(p) {
    const { formats } = p;
    switch (formats.length) {
        case 0:
            return React.createElement(ViewStringDefault, Object.assign({}, p));
        case 1: {
            const format = formats[0];
            switch (format) {
                case 'date': {
                    return (React.createElement(MyDatePicker, { path: p.path, value: p.value.value, onChange: (s) => p.dispatch({
                            tag: 'update-property',
                            path: p.path,
                            value: (0, JsonValue_1.jvString)(s),
                        }), isInvalid: isInvalid(p), invalidText: errorsToInvalidText(p), t: p.model.t, language: p.language }));
                }
                case 'date-time': {
                    const invalid = isInvalid(p);
                    const dt = new MyDateTime(p.value.value);
                    return (React.createElement("div", { className: 'date-time-picker' },
                        React.createElement("div", { className: 'date-time-picker__date' },
                            React.createElement(MyDatePicker, { path: p.path, value: dt.date, onChange: (s) => {
                                    const dt2 = dt.setDate(s);
                                    p.dispatch({
                                        tag: 'update-property',
                                        path: p.path,
                                        value: (0, JsonValue_1.jvString)(dt2.dateTime),
                                    });
                                }, isInvalid: invalid, invalidText: errorsToInvalidText(p), t: p.model.t, language: p.language })),
                        React.createElement("div", { className: 'date-time-picker__time' },
                            React.createElement(MyTimePicker, { path: p.path, onChange: (s) => {
                                    const dt2 = dt.setTime(s);
                                    p.dispatch({
                                        tag: 'update-property',
                                        path: p.path,
                                        value: (0, JsonValue_1.jvString)(dt2.dateTime),
                                    });
                                }, value: dt.time.fullTime, isInvalid: invalid, invalidText: errorsToInvalidText(p), t: p.model.t }))));
                }
                case 'time': {
                    return (React.createElement(MyTimePicker, { path: p.path, onChange: (s) => p.dispatch({
                            tag: 'update-property',
                            path: p.path,
                            value: (0, JsonValue_1.jvString)(s),
                        }), value: p.value.value, isInvalid: isInvalid(p), invalidText: errorsToInvalidText(p), t: p.model.t }));
                }
                default:
                    return React.createElement(ViewStringDefault, Object.assign({}, p));
            }
        }
        default: {
            return React.createElement(ViewStringDefault, Object.assign({}, p));
        }
    }
}
function ViewBoolean(p) {
    const { t } = p.model;
    return (React.createElement(exports.WrapErrors, Object.assign({}, p),
        React.createElement("div", { className: "checkbox-wrapper" },
            React.createElement(carbon_components_react_1.Checkbox, { labelText: t('booleanValueLabel', {
                    path: p.path.format('.'),
                }).toString(), hideLabel: true, id: 'input-' + p.path.format('_'), checked: p.value.value, disabled: p.model.adding.isJust(), onChange: () => {
                    dispatchUpdateProperty(p, {
                        tag: 'jv-boolean',
                        value: !p.value.value,
                    });
                } }))));
}
function ViewNull(p) {
    const classes = ['js-null']
        .concat(isInvalid(p) ? ['form-error'] : [])
        .join(' ');
    return (React.createElement(exports.WrapErrors, Object.assign({}, p),
        React.createElement("div", { className: classes }, "null")));
}
const onMenuTriggerClick = (dispatch, propertyPath) => (e) => {
    const btnBox = e.target;
    if (btnBox) {
        const rb = (0, tea_pop_core_1.box)((0, tea_pop_core_1.pos)(e.clientX, e.clientY), tea_pop_core_1.Dim.zero);
        dispatch({
            tag: 'menu-trigger-clicked',
            path: propertyPath,
            refBox: rb, //box(rb.p.add(pos(rb.d.w / 2, rb.d.h / 2)), Dim.zero),
        });
    }
};
exports.onMenuTriggerClick = onMenuTriggerClick;
function ViewErrors(props) {
    return props.errors.length > 0 ? (React.createElement("div", { className: "form-errors" }, props.errors.map((e) => e.message).join(', '))) : (React.createElement(React.Fragment, null));
}
exports.ViewErrors = ViewErrors;
const WrapErrors = (p) => {
    const errorsAtPath = getErrorsAtPath(p);
    return (React.createElement(React.Fragment, null,
        p.children,
        errorsAtPath.length > 0 ? React.createElement(ViewErrors, { errors: errorsAtPath }) : React.createElement(React.Fragment, null)));
};
exports.WrapErrors = WrapErrors;
function MenuTrigger(props) {
    const { disabled, path, dispatch, t, renderOptions } = props;
    return (React.createElement(carbon_components_react_1.Button, { iconDescription: t('icon.openMenu'), disabled: disabled, size: 'sm', kind: 'ghost', renderIcon: _16_4.default, tooltipPosition: 'left', hasIconOnly: true, onClick: (0, exports.onMenuTriggerClick)(dispatch, path), style: {
            overflow: (renderOptions === null || renderOptions === void 0 ? void 0 : renderOptions.hideMenuTooltip) ? 'hidden' : undefined,
        } }));
}
exports.MenuTrigger = MenuTrigger;
const allOffsets = [
    'Z',
    '+01:00',
    '+02:00',
    '+03:00',
    '+04:00',
    '+04:30',
    '+05:00',
    '+05:30',
    '+05:45',
    '+06:00',
    '+06:30',
    '+07:00',
    '+08:00',
    '+08:45',
    '+09:00',
    '+09:30',
    '+10:00',
    '+10:30',
    '+11:00',
    '+12:00',
    '+12:45',
    '+13:00',
    '+14:00',
    '-01:00',
    '-02:00',
    '-02:30',
    '-03:00',
    '-04:00',
    '-05:00',
    '-06:00',
    '-07:00',
    '-08:00',
    '-09:00',
    '-09:30',
    '-10:00',
    '-11:00',
    '-12:00',
];
class MyDateTime {
    constructor(dateTime) {
        this.dateTime = dateTime;
        this.date = '';
        this.time = new MyTime('');
        const sepIndex = dateTime.indexOf('T');
        if (sepIndex !== -1) {
            // separator found, extract date and time parts
            this.date = dateTime.substring(0, sepIndex);
            this.time = new MyTime(dateTime.substring(sepIndex + 1));
        }
    }
    setDate(date) {
        return new MyDateTime(date + 'T' + this.time.fullTime);
    }
    setTime(time) {
        return new MyDateTime(this.date + 'T' + time);
    }
}
class MyTime {
    constructor(fullTime) {
        this.fullTime = fullTime;
        this.time = fullTime;
        this.offset = '';
        for (let i = 0; i < allOffsets.length; i++) {
            const offset = allOffsets[i];
            if (fullTime.endsWith(offset)) {
                this.time = fullTime.substring(0, fullTime.length - offset.length);
                this.offset = offset;
                break;
            }
        }
    }
    setTime(time) {
        return new MyTime(time + this.offset);
    }
    setOffset(offset) {
        return new MyTime(this.time + offset);
    }
}
