// tea-pop-menu/dist/Model.js
"use strict";
/*
 * MIT License
 *
 * Copyright (c) 2020 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.keyboardNavigated = exports.menuStatePlacing = exports.initialModel = void 0;
const tea_cup_core_1 = require("tea-cup-core");
function initialModel(menu, refBox) {
    return {
        uuid: tea_cup_core_1.nothing,
        windowSize: tea_cup_core_1.nothing,
        menu,
        state: menuStatePlacing(refBox),
        error: tea_cup_core_1.nothing,
        child: tea_cup_core_1.nothing,
        navigatedWithKeyboard: false,
        subMenuCounter: 0,
    };
}
exports.initialModel = initialModel;
function menuStatePlacing(refBox) {
    return {
        tag: 'placing',
        refBox,
    };
}
exports.menuStatePlacing = menuStatePlacing;
function keyboardNavigated(model, navigatedWithKeyboard = true) {
    return Object.assign(Object.assign({}, model), { navigatedWithKeyboard });
}
exports.keyboardNavigated = keyboardNavigated;
