// tea-pop-menu/dist/Msg.js
"use strict";
/*
 * MIT License
 *
 * Copyright (c) 2020 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.docMouseDown = exports.gotItemBox = exports.childMsg = exports.gotKeyDown = exports.gotMenuBox = exports.gotUuid = exports.gotwindowBlur = exports.gotWindowDimensions = exports.noop = void 0;
function noop() {
    return { tag: 'noop' };
}
exports.noop = noop;
function gotWindowDimensions(d) {
    return {
        tag: 'got-window-dimensions',
        d,
    };
}
exports.gotWindowDimensions = gotWindowDimensions;
function gotwindowBlur() {
    return {
        tag: 'window-blur',
    };
}
exports.gotwindowBlur = gotwindowBlur;
function gotUuid(uuid) {
    return {
        tag: 'got-uuid',
        uuid,
    };
}
exports.gotUuid = gotUuid;
function gotMenuBox(r) {
    return {
        tag: 'got-menu-box',
        r,
    };
}
exports.gotMenuBox = gotMenuBox;
function gotKeyDown(key) {
    return {
        tag: 'key-down',
        key,
    };
}
exports.gotKeyDown = gotKeyDown;
function childMsg(m) {
    return {
        tag: 'child-msg',
        m,
    };
}
exports.childMsg = childMsg;
function gotItemBox(item, r, subMenuCounter, selectFirst) {
    return {
        tag: 'got-item-box',
        item,
        selectFirst,
        r,
        subMenuCounter,
    };
}
exports.gotItemBox = gotItemBox;
function docMouseDown() {
    return {
        tag: 'doc-mouse-down',
    };
}
exports.docMouseDown = docMouseDown;
