// tea-pop-menu/dist/TeaPop.js
"use strict";
/*
 * MIT License
 *
 * Copyright (c) 2020 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.subscriptions = exports.update = exports.open = void 0;
const Msg_1 = require("./Msg");
const tea_cup_core_1 = require("tea-cup-core");
const Model_1 = require("./Model");
const Menu_1 = require("./Menu");
const tea_pop_core_1 = require("tea-pop-core");
const react_tea_cup_1 = require("react-tea-cup");
const Logger_1 = require("./Logger");
function open(menu, refBox, selectFirst = false) {
    Logger_1.log('teapop', 'open', 'refBox', refBox);
    return [
        Model_1.initialModel(selectFirst ? menu.selectFirstItem() : menu.deselectAll(), refBox),
        tea_cup_core_1.Cmd.batch([
            tea_cup_core_1.Task.perform(getWindowDimensions, (d) => Msg_1.gotWindowDimensions(d)),
            tea_cup_core_1.Task.perform(tea_cup_core_1.uuid(), (u) => Msg_1.gotUuid(u)),
        ]),
    ];
}
exports.open = open;
function postOpen(model) {
    if (model.uuid.type === 'Nothing') {
        return tea_cup_core_1.noCmd(model);
    }
    if (model.windowSize.type === 'Nothing') {
        return tea_cup_core_1.noCmd(model);
    }
    const cmd = tea_cup_core_1.Task.attempt(Menu_1.menuTask(model.uuid.value).map((e) => {
        const rect = e.getBoundingClientRect();
        const b = tea_pop_core_1.Box.fromDomRect(rect);
        Logger_1.log('teapop', 'postOpen', 'menu box', b);
        return b;
    }), (x) => Msg_1.gotMenuBox(x));
    return [model, cmd];
}
function withOut(mac, outMsg = tea_cup_core_1.nothing) {
    return [mac[0], mac[1], outMsg];
}
function update(msg, model) {
    switch (msg.tag) {
        case 'got-window-dimensions': {
            if (model.state.tag === 'open') {
                return withOut(tea_cup_core_1.noCmd(model), tea_cup_core_1.just({ tag: 'request-close' }));
            }
            return withOut(postOpen(Object.assign(Object.assign({}, model), { windowSize: tea_cup_core_1.just(msg.d) })));
        }
        case 'window-blur': {
            return withOut(tea_cup_core_1.noCmd(model), tea_cup_core_1.just({ tag: 'request-close' }));
        }
        case 'got-uuid': {
            return withOut(postOpen(Object.assign(Object.assign({}, model), { uuid: tea_cup_core_1.just(msg.uuid) })));
        }
        case 'got-menu-box': {
            if (model.state.tag === 'placing') {
                const state = model.state;
                return withOut(tea_cup_core_1.noCmd(model.windowSize
                    .map((windowSize) => {
                    const newModel = msg.r.match((menuBox) => {
                        Logger_1.log('teapop', 'got-menu-box', 'menuBox', menuBox);
                        const placedBox = tea_pop_core_1.place(windowSize, state.refBox, menuBox.d);
                        Logger_1.log('teapop', 'got-menu-box', 'placedBox', placedBox);
                        return Object.assign(Object.assign({}, model), { state: {
                                tag: 'open',
                                box: placedBox,
                            } });
                    }, (err) => (Object.assign(Object.assign({}, model), { error: tea_cup_core_1.just(err) })));
                    return newModel;
                })
                    .withDefault(model)));
            }
            Logger_1.log('teapop', 'got-menu-box', 'not placing');
            return withOut(tea_cup_core_1.noCmd(model));
        }
        case 'key-down': {
            switch (msg.key) {
                case 'Escape': {
                    return withOut(collapseLastSubMenu(Model_1.keyboardNavigated(model)), model.child.isNothing() ? tea_cup_core_1.just({ tag: 'request-close' }) : tea_cup_core_1.nothing);
                }
                case 'ArrowDown':
                case 'ArrowUp': {
                    return mapLastMenu(Model_1.keyboardNavigated(model), (lastModel) => {
                        return withOut(tea_cup_core_1.noCmd(Object.assign(Object.assign({}, lastModel), { menu: lastModel.menu.moveSelection(msg.key === 'ArrowDown') })));
                    });
                }
                case 'ArrowLeft':
                    return withOut(collapseLastSubMenu(Model_1.keyboardNavigated(model)));
                case 'ArrowRight':
                    return expandLastSubMenu(Model_1.keyboardNavigated(model));
                case 'Enter':
                case ' ': {
                    return toggleOrSelectItem(model);
                }
                default:
                    return withOut(tea_cup_core_1.noCmd(model));
            }
        }
        case 'mouse-move': {
            if (model.menu.isSelected(msg.item)) {
                return withOut(tea_cup_core_1.noCmd(model));
            }
            const newModel = Object.assign(Object.assign({}, model), { subMenuCounter: model.subMenuCounter + 1 });
            if (model.navigatedWithKeyboard) {
                return withOut(tea_cup_core_1.noCmd(Model_1.keyboardNavigated(newModel, false)));
            }
            if (model.uuid.type === 'Nothing') {
                return withOut(tea_cup_core_1.noCmd(newModel));
            }
            const menu = model.menu.selectItem(msg.item);
            const uuid = model.uuid.value;
            return withOut(openSubMenu(Object.assign(Object.assign({}, newModel), { menu, child: tea_cup_core_1.nothing }), uuid, msg.item, msg.itemIndex, false));
        }
        case 'mouse-leave': {
            return withOut(tea_cup_core_1.noCmd(model.child.isJust()
                ? model
                : Object.assign(Object.assign({}, model), { menu: model.menu.deselectAll() })));
        }
        case 'got-item-box': {
            if (msg.subMenuCounter !== model.subMenuCounter) {
                return withOut(tea_cup_core_1.noCmd(model));
            }
            return withOut(msg.r.match((itemBox) => {
                const newModel = Object.assign(Object.assign({}, model), { menu: model.menu.selectItem(msg.item) });
                return msg.item.subMenu
                    .map((subMenu) => {
                    // we have a sub menu so we need
                    // to open a new Menu !
                    const mac = open(subMenu, itemBox, msg.selectFirst);
                    const newModel2 = Object.assign(Object.assign({}, newModel), { child: tea_cup_core_1.just(mac[0]) });
                    return tea_cup_core_1.Tuple.t2n(newModel2, mac[1].map(Msg_1.childMsg));
                })
                    .withDefaultSupply(() => {
                    // close any existing child !
                    return newModel.child
                        .map(() => tea_cup_core_1.noCmd(Object.assign(Object.assign({}, model), { child: tea_cup_core_1.nothing })))
                        .withDefaultSupply(() => tea_cup_core_1.noCmd(model));
                });
            }, (err) => tea_cup_core_1.noCmd(Object.assign(Object.assign({}, model), { error: tea_cup_core_1.just(err) }))));
        }
        case 'child-msg': {
            return model.child
                .map((child) => {
                const mco = update(msg.m, child);
                const newModel = Object.assign(Object.assign({}, model), { child: tea_cup_core_1.just(mco[0]) });
                return withOut(tea_cup_core_1.Tuple.t2n(newModel, mco[1].map(Msg_1.childMsg)), mco[2]);
            })
                .withDefaultSupply(() => withOut(tea_cup_core_1.noCmd(model)));
        }
        case 'item-clicked': {
            return withOut(tea_cup_core_1.noCmd(model), outItemSelected(msg.item));
        }
        case 'doc-mouse-down': {
            return withOut(tea_cup_core_1.noCmd(model), tea_cup_core_1.just({ tag: 'request-close' }));
        }
        case 'noop': {
            return withOut(tea_cup_core_1.noCmd(model));
        }
    }
}
exports.update = update;
function outItemSelected(item) {
    return tea_cup_core_1.just({ tag: 'item-selected', data: item.userData });
}
function openSubMenu(model, menuId, item, itemIndex, selectFirst) {
    if (item.subMenu.isNothing()) {
        return tea_cup_core_1.noCmd(model);
    }
    const subMenuCounter = model.subMenuCounter + 1;
    return tea_cup_core_1.Tuple.t2n(Object.assign(Object.assign({}, model), { subMenuCounter }), tea_cup_core_1.Task.attempt(Menu_1.menuItemTask(menuId, itemIndex).map((e) => {
        return tea_pop_core_1.Box.fromDomRect(e.getBoundingClientRect());
    }), (r) => Msg_1.gotItemBox(item, r, subMenuCounter, selectFirst)));
}
const windowEvents = new react_tea_cup_1.WindowEvents();
const documentEvents = new react_tea_cup_1.DocumentEvents();
function subscriptions(model) {
    return tea_cup_core_1.Sub.batch([
        windowEvents.on('resize', () => Msg_1.gotWindowDimensions(tea_pop_core_1.dim(window.innerWidth, window.innerHeight))),
        windowEvents.on('blur', () => Msg_1.gotwindowBlur()),
        documentEvents.on('mousedown', (evt) => {
            if (evt.button === 2) {
                return Msg_1.noop();
            }
            let t = evt.target;
            while (t) {
                // move up and try to find if we are inside a tea-pop Menu !
                if (t.classList.contains('tm')) {
                    return Msg_1.noop();
                }
                t = t.parentElement;
            }
            return Msg_1.docMouseDown();
        }, true),
        model.state.tag === 'open'
            ? documentEvents.on('keydown', (e) => Msg_1.gotKeyDown(e.key))
            : tea_cup_core_1.Sub.none(),
    ]);
}
exports.subscriptions = subscriptions;
const getWindowDimensions = tea_cup_core_1.Task.succeedLazy(() => tea_pop_core_1.dim(window.innerWidth, window.innerHeight));
function mapLastMenu(model, f) {
    switch (model.child.type) {
        case 'Nothing': {
            // I'm the last model !
            return f(model);
        }
        case 'Just': {
            const mac = mapLastMenu(model.child.value, f);
            return [Object.assign(Object.assign({}, model), { child: tea_cup_core_1.just(mac[0]) }), mac[1].map(Msg_1.childMsg), mac[2]];
        }
    }
}
function collapseLastSubMenu(model) {
    switch (model.child.type) {
        case 'Nothing':
            return tea_cup_core_1.noCmd(model);
        case 'Just': {
            // I have a child : close it if he has no child !
            const child = model.child.value;
            if (child.child.isJust()) {
                return tea_cup_core_1.Tuple.fromNative(collapseLastSubMenu(child))
                    .mapFirst((c) => (Object.assign(Object.assign({}, model), { child: tea_cup_core_1.just(c) })))
                    .mapSecond((cmd) => cmd.map(Msg_1.childMsg))
                    .toNative();
            }
            else {
                return tea_cup_core_1.noCmd(Object.assign(Object.assign({}, model), { child: tea_cup_core_1.nothing }));
            }
        }
    }
}
function expandLastSubMenu(model) {
    return mapLastMenu(model, (lastModel) => {
        return lastModel.menu.selectedItem
            .map((selectedItem) => {
            return selectedItem.subMenu
                .map(() => {
                if (lastModel.uuid.type === 'Nothing') {
                    return withOut(tea_cup_core_1.noCmd(lastModel));
                }
                const uuid = lastModel.uuid.value;
                const mac = lastModel.menu
                    .indexOfItem(selectedItem)
                    .map((itemIndex) => {
                    return openSubMenu(lastModel, uuid, selectedItem, itemIndex, true);
                })
                    .withDefaultSupply(() => tea_cup_core_1.noCmd(lastModel));
                return withOut(mac);
            })
                .withDefaultSupply(() => {
                return withOut(tea_cup_core_1.noCmd(lastModel));
            });
        })
            .withDefaultSupply(() => {
            return withOut(tea_cup_core_1.noCmd(lastModel));
        });
    });
}
function toggleOrSelectItem(model) {
    return mapLastMenu(model, (lastModel) => {
        return lastModel.menu.selectedItem
            .map((selectedItem) => {
            if (lastModel.uuid.type === 'Nothing') {
                return withOut(tea_cup_core_1.noCmd(lastModel));
            }
            const uuid = lastModel.uuid.value;
            // do we have a sub-menu ?
            return selectedItem.subMenu
                .map(() => {
                // we have a sub-menu, expand it
                const mac = lastModel.menu
                    .indexOfItem(selectedItem)
                    .map((itemIndex) => openSubMenu(lastModel, uuid, selectedItem, itemIndex, true))
                    .withDefaultSupply(() => tea_cup_core_1.noCmd(lastModel));
                return withOut(mac);
            })
                .withDefaultSupply(() => {
                // no sub-menu, select the item
                return withOut(tea_cup_core_1.noCmd(lastModel), outItemSelected(selectedItem));
            });
        })
            .withDefaultSupply(() => withOut(tea_cup_core_1.noCmd(lastModel)));
    });
}
