// tea-pop-menu/dist/ViewMenu.js
"use strict";
/*
 * MIT License
 *
 * Copyright (c) 2020 Rémi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewMenuItem = exports.ViewMenu = void 0;
const React = __importStar(require("react"));
const Menu_1 = require("./Menu");
const Msg_1 = require("./Msg");
const tea_cup_core_1 = require("tea-cup-core");
const tea_pop_core_1 = require("tea-pop-core");
function ViewMenu(props) {
    const { model, dispatch, renderer } = props;
    const { menu, state, uuid, windowSize } = model;
    if (uuid.type === 'Nothing') {
        return React.createElement(React.Fragment, null);
    }
    if (windowSize.type === 'Nothing') {
        return React.createElement(React.Fragment, null);
    }
    const renderItems = () => menu.elems.map((element, index) => {
        switch (element.tag) {
            case 'item': {
                return (React.createElement(ViewMenuItem, Object.assign({ key: `item-${index}`, uuid: uuid.value, itemIndex: index, menu: menu, item: element }, props)));
            }
            case 'separator': {
                return React.createElement("div", { key: `sep-${index}`, className: "tm-separator" });
            }
        }
    });
    switch (state.tag) {
        case 'placing': {
            const { refBox } = state;
            return (React.createElement("div", { className: "tm-placer", style: {
                    position: 'absolute',
                    top: 0,
                    left: 0,
                    bottom: 0,
                    right: 0,
                    overflow: 'hidden',
                }, onContextMenu: tea_pop_core_1.stopEvent },
                React.createElement("div", { className: "tm", id: Menu_1.menuId(uuid.value), style: {
                        position: 'absolute',
                        top: refBox.p.y,
                        left: refBox.p.x,
                        visibility: 'hidden',
                    }, onContextMenu: tea_pop_core_1.stopEvent }, renderItems())));
        }
        case 'open': {
            const { box } = state;
            const { p, d } = box;
            return (React.createElement(React.Fragment, null,
                React.createElement("div", { className: "tm", id: Menu_1.menuId(uuid.value), style: {
                        position: 'fixed',
                        top: p.y,
                        left: p.x,
                        width: d.w,
                        height: d.h,
                    }, onContextMenu: tea_pop_core_1.stopEvent }, renderItems()),
                model.child
                    .map((child) => (React.createElement(ViewMenu, { model: child, dispatch: tea_cup_core_1.map(dispatch, Msg_1.childMsg), renderer: renderer })))
                    .withDefaultSupply(() => (React.createElement(React.Fragment, null)))));
        }
    }
}
exports.ViewMenu = ViewMenu;
function ViewMenuItem(props) {
    const { menu, item, renderer, dispatch, uuid, itemIndex } = props;
    const selected = menu.isSelected(item);
    return (React.createElement("div", { id: Menu_1.menuItemId(uuid, itemIndex), onMouseMove: () => dispatch({ tag: 'mouse-move', item, itemIndex }), onMouseLeave: () => dispatch({ tag: 'mouse-leave', item, itemIndex }), onClick: () => {
            dispatch({ tag: 'item-clicked', item });
        } }, renderer({
        data: item.userData,
        active: selected,
        hasSubMenu: item.subMenu.isJust(),
    })));
}
exports.ViewMenuItem = ViewMenuItem;
